// @nuvem/primeng-components/esm2015/lib/datatable/show-column/show-column.module.js
import { NgModule, CUSTOM_ELEMENTS_SCHEMA } from '@angular/core';
import { CommonModule } from '@angular/common';
import { ButtonModule } from 'primeng/button';
import { DropdownModule } from 'primeng/dropdown';
import { FormsModule, ReactiveFormsModule } from '@angular/forms';
import { CheckboxModule } from 'primeng/checkbox';
import { MultiSelectModule } from 'primeng/multiselect';
import { InputTextModule } from 'primeng/inputtext';
import { ShowColumn } from './show-column.component';
import * as ɵngcc0 from '@angular/core';
export class ShowColumnModule {
}
ShowColumnModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: ShowColumnModule });
ShowColumnModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function ShowColumnModule_Factory(t) { return new (t || ShowColumnModule)(); }, imports: [[
            CommonModule,
            MultiSelectModule,
            ButtonModule,
            CheckboxModule,
            DropdownModule,
            FormsModule,
            ReactiveFormsModule,
            InputTextModule
        ]] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(ShowColumnModule, { declarations: function () { return [ShowColumn]; }, imports: function () { return [CommonModule,
        MultiSelectModule,
        ButtonModule,
        CheckboxModule,
        DropdownModule,
        FormsModule,
        ReactiveFormsModule,
        InputTextModule]; }, exports: function () { return [ShowColumn]; } }); })();
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(ShowColumnModule, [{
        type: NgModule,
        args: [{
                imports: [
                    CommonModule,
                    MultiSelectModule,
                    ButtonModule,
                    CheckboxModule,
                    DropdownModule,
                    FormsModule,
                    ReactiveFormsModule,
                    InputTextModule
                ],
                declarations: [
                    ShowColumn
                ],
                exports: [
                    ShowColumn
                ],
                schemas: [CUSTOM_ELEMENTS_SCHEMA]
            }]
    }], null, null); })();
