// @nuvem/primeng-components/esm2015/lib/form-notification/form-notification.module.js
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { FormNotificationBaseModule } from '@nuvem/angular-base';
import { FormNotificationComponent } from './form-notification.component';
/**
 * Módulo para adicionar funcionalidade de notificação de erros de formulários.
 * @module
 */
import * as ɵngcc0 from '@angular/core';
export class FormNotificationModule {
}
FormNotificationModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: FormNotificationModule });
FormNotificationModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function FormNotificationModule_Factory(t) { return new (t || FormNotificationModule)(); }, imports: [[
            CommonModule,
            FormNotificationBaseModule
        ]] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(FormNotificationModule, { declarations: function () { return [FormNotificationComponent]; }, imports: function () { return [CommonModule,
        FormNotificationBaseModule]; }, exports: function () { return [FormNotificationComponent]; } }); })();
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(FormNotificationModule, [{
        type: NgModule,
        args: [{
                imports: [
                    CommonModule,
                    FormNotificationBaseModule
                ],
                declarations: [
                    FormNotificationComponent
                ],
                exports: [
                    FormNotificationComponent
                ]
            }]
    }], null, null); })();
