// @nuvem/primeng-components/esm2015/lib/form-notification/form-notification.component.js
import { Component } from "@angular/core";
import { FormNotificationComponentSuperclass } from '@nuvem/angular-base';
import * as ɵngcc0 from '@angular/core';

const _c0 = ["form-notification", ""];
const _c1 = ["*"];
export class FormNotificationComponent extends FormNotificationComponentSuperclass {
}
FormNotificationComponent.ɵfac = function FormNotificationComponent_Factory(t) { return ɵFormNotificationComponent_BaseFactory(t || FormNotificationComponent); };
FormNotificationComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: FormNotificationComponent, selectors: [["", "form-notification", ""]], features: [ɵngcc0.ɵɵInheritDefinitionFeature], attrs: _c0, ngContentSelectors: _c1, decls: 3, vars: 1, consts: [[1, "p-invalid"]], template: function FormNotificationComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵprojectionDef();
        ɵngcc0.ɵɵprojection(0);
        ɵngcc0.ɵɵelementStart(1, "small", 0);
        ɵngcc0.ɵɵtext(2);
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵtextInterpolate(ctx.message);
    } }, encapsulation: 2 });
const ɵFormNotificationComponent_BaseFactory = /*@__PURE__*/ ɵngcc0.ɵɵgetInheritedFactory(FormNotificationComponent);
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(FormNotificationComponent, [{
        type: Component,
        args: [{
                selector: '[form-notification]',
                template: "<ng-content></ng-content>\n<small class=\"p-invalid\">{{message}}</small>"
            }]
    }], null, null); })();
