// @nuvem/primeng-components/esm2015/lib/extra-button/datatable-splitbutton.module.js
import { DatatableSplitbuttonComponent } from './datatable-splitbutton.component';
import { NgModule, CUSTOM_ELEMENTS_SCHEMA } from '@angular/core';
import { SplitButtonModule } from 'primeng/splitbutton';
/**
 * Class DatatableSplitbuttonModule
 * @class
 */
import * as ɵngcc0 from '@angular/core';
export class DatatableSplitbuttonModule {
}
DatatableSplitbuttonModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: DatatableSplitbuttonModule });
DatatableSplitbuttonModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function DatatableSplitbuttonModule_Factory(t) { return new (t || DatatableSplitbuttonModule)(); }, providers: [], imports: [[
            SplitButtonModule
        ], SplitButtonModule] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(DatatableSplitbuttonModule, { declarations: function () { return [DatatableSplitbuttonComponent]; }, imports: function () { return [SplitButtonModule]; }, exports: function () { return [DatatableSplitbuttonComponent,
        SplitButtonModule]; } }); })();
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(DatatableSplitbuttonModule, [{
        type: NgModule,
        args: [{
                imports: [
                    SplitButtonModule
                ],
                declarations: [
                    DatatableSplitbuttonComponent
                ],
                providers: [],
                exports: [
                    DatatableSplitbuttonComponent,
                    SplitButtonModule
                ],
                schemas: [CUSTOM_ELEMENTS_SCHEMA]
            }]
    }], null, null); })();
