// @nuvem/angular-base/esm2015/lib/form-notification/components/form-notification-superclass.component.js
import { FormNotificationService } from '../services/form-notification.service';
import { Directive, Input } from "@angular/core";
import { FieldFormNotification } from '../types/field-form-notification.type';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '../services/form-notification.service';
export class FormNotificationComponentSuperclass {
    /**
     * Método construtor responsável por carregar o serviço de validação
     * @param formNotificationService
     */
    constructor(formNotificationService) {
        this.formNotificationService = formNotificationService;
    }
    /**
     * Método Inicial para subscrição no endpoint responsável por validar o formulário
     */
    ngOnInit() {
        this.subscription = this.formNotificationService
            .notifications(this.form, this.name)
            .subscribe((notification) => {
            if (notification instanceof FieldFormNotification) {
                this.message = notification.message;
            }
            else {
                this.message = null;
            }
        });
    }
    /**
     * Método responsável por realizar a saida das subscrições do endpont de validação
     */
    ngOnDestroy() {
        this.subscription.unsubscribe();
    }
}
FormNotificationComponentSuperclass.ɵfac = function FormNotificationComponentSuperclass_Factory(t) { return new (t || FormNotificationComponentSuperclass)(ɵngcc0.ɵɵdirectiveInject(ɵngcc1.FormNotificationService)); };
FormNotificationComponentSuperclass.ɵdir = ɵngcc0.ɵɵdefineDirective({ type: FormNotificationComponentSuperclass, selectors: [["form-notification-component-superclass"]], inputs: { name: "name", form: "form" } });
FormNotificationComponentSuperclass.ctorParameters = () => [
    { type: FormNotificationService }
];
FormNotificationComponentSuperclass.propDecorators = {
    name: [{ type: Input }],
    form: [{ type: Input }]
};
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(FormNotificationComponentSuperclass, [{
        type: Directive,
        args: [{
                selector: 'form-notification-component-superclass'
            }]
    }], function () { return [{ type: ɵngcc1.FormNotificationService }]; }, { name: [{
            type: Input
        }], form: [{
            type: Input
        }] }); })();
