// @nuvem/primeng-components/esm2015/lib/menu/menu.module.js
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { MenuComponent } from './menu.component';
import { SubMenuComponent } from './submenu.component';
import { RouterModule } from '@angular/router';
import { MenusService } from './menu.service';
import { TooltipModule } from 'primeng/tooltip';
import * as ɵngcc0 from '@angular/core';
export class MenuModule {
}
MenuModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: MenuModule });
MenuModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function MenuModule_Factory(t) { return new (t || MenuModule)(); }, providers: [
        MenusService
    ], imports: [[
            CommonModule,
            RouterModule,
            TooltipModule
        ]] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(MenuModule, { declarations: function () { return [MenuComponent,
        SubMenuComponent]; }, imports: function () { return [CommonModule,
        RouterModule,
        TooltipModule]; }, exports: function () { return [MenuComponent,
        SubMenuComponent]; } }); })();
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(MenuModule, [{
        type: NgModule,
        args: [{
                declarations: [
                    MenuComponent,
                    SubMenuComponent
                ],
                imports: [
                    CommonModule,
                    RouterModule,
                    TooltipModule
                ],
                exports: [
                    MenuComponent,
                    SubMenuComponent
                ],
                providers: [
                    MenusService
                ]
            }]
    }], null, null); })();
