// @nuvem/primeng-components/esm2015/lib/menu/menu.component.js
import { Component } from '@angular/core';
import { MenusService } from './menu.service';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from './menu.service';
import * as ɵngcc2 from '@angular/common';
import * as ɵngcc3 from './submenu.component';

function MenuComponent_li_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "li", 2);
} if (rf & 2) {
    const item_r1 = ctx.$implicit;
    const i_r2 = ctx.index;
    ɵngcc0.ɵɵproperty("item", item_r1)("index", i_r2);
} }
export class MenuComponent {
    constructor(menuService) {
        this.menuService = menuService;
    }
}
MenuComponent.ɵfac = function MenuComponent_Factory(t) { return new (t || MenuComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc1.MenusService)); };
MenuComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: MenuComponent, selectors: [["app-menu"]], decls: 2, vars: 1, consts: [[1, "layout-menu"], ["submenu", "", "root", "true", 3, "item", "index", 4, "ngFor", "ngForOf"], ["submenu", "", "root", "true", 3, "item", "index"]], template: function MenuComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelementStart(0, "ul", 0);
        ɵngcc0.ɵɵtemplate(1, MenuComponent_li_1_Template, 1, 2, "li", 1);
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngForOf", ctx.menuService.itens);
    } }, directives: [ɵngcc2.NgForOf, ɵngcc3.SubMenuComponent], encapsulation: 2 });
MenuComponent.ctorParameters = () => [
    { type: MenusService }
];
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(MenuComponent, [{
        type: Component,
        args: [{
                selector: 'app-menu',
                template: "<ul class=\"layout-menu\">\n    <li submenu *ngFor=\"let item of menuService.itens; let i = index;\" [item]=\"item\" [index]=\"i\" root=\"true\"></li>\n</ul>"
            }]
    }], function () { return [{ type: ɵngcc1.MenusService }]; }, null); })();
