// @nuvem/primeng-components/esm2015/lib/menu/submenu.component.js
import { Component, Input } from '@angular/core';
import { trigger, state, style, transition, animate } from '@angular/animations';
import { MenusService } from './menu.service';
import { NavigationEnd, Router } from '@angular/router';
import { filter } from 'rxjs/operators';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from './menu.service';
import * as ɵngcc2 from '@angular/router';
import * as ɵngcc3 from '@angular/common';
import * as ɵngcc4 from 'primeng/tooltip';

const _c0 = ["submenu", ""];
function SubMenuComponent_div_0_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div");
    ɵngcc0.ɵɵelementStart(1, "span", 3);
    ɵngcc0.ɵɵtext(2);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r0 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵtextInterpolate(ctx_r0.item.label);
} }
function SubMenuComponent_a_1_span_4_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "span", 8);
    ɵngcc0.ɵɵtext(1);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r4 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵproperty("ngClass", ctx_r4.item.badgeClass);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵtextInterpolate(ctx_r4.item.badge);
} }
function SubMenuComponent_a_1_i_5_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "i", 9);
} }
function SubMenuComponent_a_1_Template(rf, ctx) { if (rf & 1) {
    const _r7 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "a", 4);
    ɵngcc0.ɵɵlistener("click", function SubMenuComponent_a_1_Template_a_click_0_listener($event) { ɵngcc0.ɵɵrestoreView(_r7); const ctx_r6 = ɵngcc0.ɵɵnextContext(); return ctx_r6.itemClick($event); })("keydown.enter", function SubMenuComponent_a_1_Template_a_keydown_enter_0_listener($event) { ɵngcc0.ɵɵrestoreView(_r7); const ctx_r8 = ɵngcc0.ɵɵnextContext(); return ctx_r8.itemClick($event); })("mouseenter", function SubMenuComponent_a_1_Template_a_mouseenter_0_listener() { ɵngcc0.ɵɵrestoreView(_r7); const ctx_r9 = ɵngcc0.ɵɵnextContext(); return ctx_r9.onMouseEnter(); });
    ɵngcc0.ɵɵelement(1, "i", 5);
    ɵngcc0.ɵɵelementStart(2, "span", 3);
    ɵngcc0.ɵɵtext(3);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵtemplate(4, SubMenuComponent_a_1_span_4_Template, 2, 2, "span", 6);
    ɵngcc0.ɵɵtemplate(5, SubMenuComponent_a_1_i_5_Template, 1, 0, "i", 7);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r1 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵproperty("ngClass", ctx_r1.item.class)("pTooltip", ctx_r1.item.label)("tooltipDisabled", ctx_r1.active || !(ctx_r1.root && ctx_r1.menuService.isSlim() && !ctx_r1.menuService.isMobile));
    ɵngcc0.ɵɵattribute("href", ctx_r1.item.url, ɵngcc0.ɵɵsanitizeUrl)("target", ctx_r1.item.target)("tabindex", 0);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngClass", ctx_r1.item.icon);
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵtextInterpolate(ctx_r1.item.label);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r1.item.badge && !ctx_r1.root);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r1.item.items);
} }
function SubMenuComponent_a_2_span_4_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "span", 8);
    ɵngcc0.ɵɵtext(1);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r10 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵproperty("ngClass", ctx_r10.item.badgeClass);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵtextInterpolate(ctx_r10.item.badge);
} }
function SubMenuComponent_a_2_i_5_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "i", 9);
} }
const _c1 = function () { return { exact: true }; };
function SubMenuComponent_a_2_Template(rf, ctx) { if (rf & 1) {
    const _r13 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "a", 10);
    ɵngcc0.ɵɵlistener("click", function SubMenuComponent_a_2_Template_a_click_0_listener($event) { ɵngcc0.ɵɵrestoreView(_r13); const ctx_r12 = ɵngcc0.ɵɵnextContext(); return ctx_r12.itemClick($event); })("mouseenter", function SubMenuComponent_a_2_Template_a_mouseenter_0_listener() { ɵngcc0.ɵɵrestoreView(_r13); const ctx_r14 = ɵngcc0.ɵɵnextContext(); return ctx_r14.onMouseEnter(); });
    ɵngcc0.ɵɵelement(1, "i", 5);
    ɵngcc0.ɵɵelementStart(2, "span", 3);
    ɵngcc0.ɵɵtext(3);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵtemplate(4, SubMenuComponent_a_2_span_4_Template, 2, 2, "span", 6);
    ɵngcc0.ɵɵtemplate(5, SubMenuComponent_a_2_i_5_Template, 1, 0, "i", 7);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r2 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵproperty("routerLink", ctx_r2.item.routerLink)("routerLinkActiveOptions", ɵngcc0.ɵɵpureFunction0(11, _c1))("ngClass", ctx_r2.item.class)("pTooltip", ctx_r2.item.label)("tooltipDisabled", ctx_r2.active || !(ctx_r2.root && ctx_r2.menuService.isSlim() && !ctx_r2.menuService.isMobile));
    ɵngcc0.ɵɵattribute("target", ctx_r2.item.target)("tabindex", 0);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngClass", ctx_r2.item.icon);
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵtextInterpolate(ctx_r2.item.label);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r2.item.badge && !ctx_r2.root);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r2.item.items);
} }
function SubMenuComponent_ul_3_ng_template_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "li", 12);
} if (rf & 2) {
    const child_r16 = ctx.$implicit;
    const i_r17 = ctx.index;
    const ctx_r15 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵclassMap(child_r16.badgeClass);
    ɵngcc0.ɵɵproperty("item", child_r16)("index", i_r17)("parentKey", ctx_r15.key);
} }
function SubMenuComponent_ul_3_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "ul");
    ɵngcc0.ɵɵtemplate(1, SubMenuComponent_ul_3_ng_template_1_Template, 1, 5, "ng-template", 11);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r3 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵproperty("@children", ctx_r3.root ? "visible" : ctx_r3.active ? "visibleAnimated" : "hiddenAnimated");
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngForOf", ctx_r3.item.items);
} }
export class SubMenuComponent {
    constructor(menuService, router) {
        this.menuService = menuService;
        this.router = router;
        this.active = false;
        this.menuSourceSubscription = this.menuService.menuSource$.subscribe(key => {
            // deactivate current active menu
            if (this.active && this.key !== key && key.indexOf(this.key) !== 0) {
                this.active = false;
            }
        });
        this.menuResetSubscription = this.menuService.resetSource$.subscribe(() => {
            this.active = false;
        });
        this.router.events.pipe(filter(event => event instanceof NavigationEnd))
            .subscribe(params => {
            if (this.menuService.isHorizontal() || this.menuService.isSlim()) {
                this.active = false;
            }
            else {
                if (this.item.routerLink) {
                    this.updateActiveStateFromRoute();
                }
                else {
                    this.active = false;
                }
            }
        });
    }
    updateActiveStateFromRoute() {
        this.active = this.router.isActive(this.item.routerLink[0], this.item.items ? false : true);
    }
    /**
     * Alterar o menu root entre aberto e fechado
     */
    rootMenuToggle() {
        if (this.root) {
            this.menuService.menuHoverActive = !this.menuService.menuHoverActive;
        }
    }
    /**
     * Activa o menu por index e desativa o item ativo
     * @param index
     */
    activeMenuByIndex(index) {
        this.activeIndex = (this.activeIndex === index) ? null : index;
        return this.activeIndex;
    }
    /**
     * Verifica se foi informado um comando para o item de menu e executa caso necessário
     * @param item menu
     */
    execCommand(event, item) {
        if (item.command) {
            item.command({ 'originalEvent': event, 'item': item });
            return true;
        }
        return false;
    }
    /**
     * Esconte o menu caso este não possua itens
     * @param item menu
     */
    hideMenu(item) {
        if (!item.items) {
            if (this.menuService.isHorizontal() || this.menuService.isSlim()) {
                this.menuService.resetMenu = true;
            }
            else {
                this.menuService.resetMenu = false;
            }
            this.menuService.overlayMenuActive = false;
            this.menuService.staticMenuMobileActive = false;
            this.menuService.menuHoverActive = !this.menuService.menuHoverActive;
            return true;
        }
        return false;
    }
    itemClick(event) {
        // avoid processing disabled items
        if (this.item.disabled) {
            event.preventDefault();
            return true;
        }
        // navigate with hover in horizontal mode
        if (this.root) {
            this.menuService.menuHoverActive = !this.menuService.menuHoverActive;
        }
        // notify other items
        this.menuService.onMenuStateChange(this.key);
        // execute command
        if (this.item.command) {
            this.item.command({ originalEvent: event, item: this.item });
        }
        // toggle active state
        if (this.item.items) {
            this.active = !this.active;
        }
        else {
            // activate item
            this.active = true;
            // reset horizontal and slim menu
            if (this.menuService.isHorizontal() || this.menuService.isSlim()) {
                this.menuService.reset();
                this.menuService.menuHoverActive = false;
            }
            if (!this.menuService.isStatic()) {
                this.active = false;
            }
        }
        this.removeActiveInk(event);
    }
    onMouseEnter() {
        // activate item on hover
        if (this.root) {
            this.active = true;
        }
    }
    isActive(index) {
        return this.activeIndex === index;
    }
    get reset() {
        return this._reset;
    }
    set reset(val) {
        this._reset = val;
        if (this._reset && (this.menuService.isHorizontal() || this.menuService.isSlim())) {
            this.activeIndex = null;
        }
    }
    get parentActive() {
        return this._parentActive;
    }
    set parentActive(val) {
        this._parentActive = val;
        if (!this._parentActive) {
            this.activeIndex = null;
        }
    }
    removeActiveInk(event) {
        let currentTarget = event.currentTarget;
        setTimeout(() => {
            if (currentTarget) {
                let activeInk = currentTarget.querySelector('.p-ink-active');
                if (activeInk) {
                    if (activeInk.classList)
                        activeInk.classList.remove('p-ink-active');
                    else
                        activeInk.className = activeInk.className.replace(new RegExp('(^|\\b)' + 'p-ink-active'.split(' ').join('|') + '(\\b|$)', 'gi'), ' ');
                }
            }
        }, 401);
    }
    ngOnDestroy() {
        if (this.menuSourceSubscription) {
            this.menuSourceSubscription.unsubscribe();
        }
        if (this.menuResetSubscription) {
            this.menuResetSubscription.unsubscribe();
        }
    }
}
SubMenuComponent.ɵfac = function SubMenuComponent_Factory(t) { return new (t || SubMenuComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc1.MenusService), ɵngcc0.ɵɵdirectiveInject(ɵngcc2.Router)); };
SubMenuComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: SubMenuComponent, selectors: [["", "submenu", ""]], hostVars: 4, hostBindings: function SubMenuComponent_HostBindings(rf, ctx) { if (rf & 2) {
        ɵngcc0.ɵɵclassProp("layout-root-menuitem", ctx.root || ctx.active)("active-menuitem", ctx.active);
    } }, inputs: { reset: "reset", parentActive: "parentActive", item: "item", root: "root", visible: "visible", parentKey: "parentKey", index: "index" }, attrs: _c0, decls: 4, vars: 4, consts: [[4, "ngIf"], ["pRipple", "", 3, "ngClass", "pTooltip", "tooltipDisabled", "click", "keydown.enter", "mouseenter", 4, "ngIf"], ["routerLinkActive", "active-menuitem-routerlink", "pRipple", "", 3, "routerLink", "routerLinkActiveOptions", "ngClass", "pTooltip", "tooltipDisabled", "click", "mouseenter", 4, "ngIf"], [1, "layout-menuitem-text"], ["pRipple", "", 3, "ngClass", "pTooltip", "tooltipDisabled", "click", "keydown.enter", "mouseenter"], [1, "layout-menuitem-icon", 3, "ngClass"], ["class", "p-badge p-component p-badge-no-gutter", 3, "ngClass", 4, "ngIf"], ["class", "pi pi-fw pi-angle-down layout-submenu-toggler", 4, "ngIf"], [1, "p-badge", "p-component", "p-badge-no-gutter", 3, "ngClass"], [1, "pi", "pi-fw", "pi-angle-down", "layout-submenu-toggler"], ["routerLinkActive", "active-menuitem-routerlink", "pRipple", "", 3, "routerLink", "routerLinkActiveOptions", "ngClass", "pTooltip", "tooltipDisabled", "click", "mouseenter"], ["ngFor", "", 3, "ngForOf"], ["submenu", "", 3, "item", "index", "parentKey"]], template: function SubMenuComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵtemplate(0, SubMenuComponent_div_0_Template, 3, 1, "div", 0);
        ɵngcc0.ɵɵtemplate(1, SubMenuComponent_a_1_Template, 6, 10, "a", 1);
        ɵngcc0.ɵɵtemplate(2, SubMenuComponent_a_2_Template, 6, 12, "a", 2);
        ɵngcc0.ɵɵtemplate(3, SubMenuComponent_ul_3_Template, 2, 2, "ul", 0);
    } if (rf & 2) {
        ɵngcc0.ɵɵproperty("ngIf", ctx.root);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", !ctx.item.routerLink || ctx.item.items);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", ctx.item.routerLink && !ctx.item.items);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", ctx.item.items && ctx.root || ctx.item.items && ctx.active);
    } }, directives: [ɵngcc3.NgIf, ɵngcc3.NgClass, ɵngcc4.Tooltip, ɵngcc2.RouterLinkWithHref, ɵngcc2.RouterLinkActive, ɵngcc3.NgForOf, SubMenuComponent], encapsulation: 2, data: { animation: [
            trigger('children', [
                state('void', style({
                    height: '0px',
                    padding: '0px'
                })),
                state('hiddenAnimated', style({
                    height: '0px',
                    padding: '0px'
                })),
                state('visibleAnimated', style({
                    height: '*'
                })),
                state('visible', style({
                    height: '*'
                })),
                state('hidden', style({
                    height: '0px',
                    padding: '0px'
                })),
                transition('visibleAnimated => hiddenAnimated', animate('400ms cubic-bezier(0.86, 0, 0.07, 1)')),
                transition('hiddenAnimated => visibleAnimated', animate('400ms cubic-bezier(0.86, 0, 0.07, 1)')),
                transition('void => visibleAnimated, visibleAnimated => void', animate('400ms cubic-bezier(0.86, 0, 0.07, 1)'))
            ])
        ] } });
SubMenuComponent.ctorParameters = () => [
    { type: MenusService },
    { type: Router }
];
SubMenuComponent.propDecorators = {
    item: [{ type: Input }],
    root: [{ type: Input }],
    visible: [{ type: Input }],
    parentKey: [{ type: Input }],
    index: [{ type: Input }],
    reset: [{ type: Input }],
    parentActive: [{ type: Input }]
};
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(SubMenuComponent, [{
        type: Component,
        args: [{
                selector: '[submenu]',
                template: "<div *ngIf=\"root\">\n    <span class=\"layout-menuitem-text\">{{item.label}}</span>\n</div>\n<a [attr.href]=\"item.url\" (click)=\"itemClick($event)\" *ngIf=\"!item.routerLink || item.items\"\n    (keydown.enter)=\"itemClick($event)\" [attr.target]=\"item.target\" [attr.tabindex]=\"0\" [ngClass]=\"item.class\"\n    (mouseenter)=\"onMouseEnter()\" pRipple [pTooltip]=\"item.label\"\n    [tooltipDisabled]=\"active || !(root && menuService.isSlim() && !menuService.isMobile)\">\n    <i [ngClass]=\"item.icon\" class=\"layout-menuitem-icon\"></i>\n    <span class=\"layout-menuitem-text\">{{item.label}}</span>\n    <span class=\"p-badge p-component p-badge-no-gutter\" [ngClass]=\"item.badgeClass\"\n        *ngIf=\"item.badge && !root\">{{item.badge}}</span>\n    <i class=\"pi pi-fw pi-angle-down layout-submenu-toggler\" *ngIf=\"item.items\"></i>\n</a>\n<a (click)=\"itemClick($event)\" *ngIf=\"item.routerLink && !item.items\" [routerLink]=\"item.routerLink\"\n    routerLinkActive=\"active-menuitem-routerlink\" [routerLinkActiveOptions]=\"{exact: true}\" [attr.target]=\"item.target\"\n    [attr.tabindex]=\"0\" [ngClass]=\"item.class\" (mouseenter)=\"onMouseEnter()\" pRipple [pTooltip]=\"item.label\"\n    [tooltipDisabled]=\"active || !(root && menuService.isSlim() && !menuService.isMobile)\">\n    <i [ngClass]=\"item.icon\" class=\"layout-menuitem-icon\"></i>\n    <span class=\"layout-menuitem-text\">{{item.label}}</span>\n    <span class=\"p-badge p-component p-badge-no-gutter\" [ngClass]=\"item.badgeClass\"\n        *ngIf=\"item.badge && !root\">{{item.badge}}</span>\n    <i class=\"pi pi-fw pi-angle-down layout-submenu-toggler\" *ngIf=\"item.items\"></i>\n</a>\n<ul *ngIf=\"(item.items && root) || (item.items && active)\"\n    [@children]=\"root ? 'visible' : active ? 'visibleAnimated' : 'hiddenAnimated'\">\n    <ng-template ngFor let-child let-i=\"index\" [ngForOf]=\"item.items\">\n        <li submenu [item]=\"child\" [index]=\"i\" [parentKey]=\"key\" [class]=\"child.badgeClass\"></li>\n    </ng-template>\n</ul>",
                host: {
                    '[class.layout-root-menuitem]': 'root || active',
                    '[class.active-menuitem]': '(active)'
                },
                animations: [
                    trigger('children', [
                        state('void', style({
                            height: '0px',
                            padding: '0px'
                        })),
                        state('hiddenAnimated', style({
                            height: '0px',
                            padding: '0px'
                        })),
                        state('visibleAnimated', style({
                            height: '*'
                        })),
                        state('visible', style({
                            height: '*'
                        })),
                        state('hidden', style({
                            height: '0px',
                            padding: '0px'
                        })),
                        transition('visibleAnimated => hiddenAnimated', animate('400ms cubic-bezier(0.86, 0, 0.07, 1)')),
                        transition('hiddenAnimated => visibleAnimated', animate('400ms cubic-bezier(0.86, 0, 0.07, 1)')),
                        transition('void => visibleAnimated, visibleAnimated => void', animate('400ms cubic-bezier(0.86, 0, 0.07, 1)'))
                    ])
                ]
            }]
    }], function () { return [{ type: ɵngcc1.MenusService }, { type: ɵngcc2.Router }]; }, { reset: [{
            type: Input
        }], parentActive: [{
            type: Input
        }], item: [{
            type: Input
        }], root: [{
            type: Input
        }], visible: [{
            type: Input
        }], parentKey: [{
            type: Input
        }], index: [{
            type: Input
        }] }); })();
