// @nuvem/primeng-components/esm2015/lib/crud/crud-form.component.js
import { Component, Input } from "@angular/core";
import { ActivatedRoute } from "@angular/router";
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '@angular/router';
import * as ɵngcc2 from './form.component';
import * as ɵngcc3 from 'primeng/button';

const _c0 = function (a0) { return [a0]; };
const _c1 = ["*"];
export class CrudFormComponent {
    constructor(route) {
        this.route = route;
        this.disableSave = false;
        this.listRoute = "../../list";
    }
    ngOnInit() {
        this.route.paramMap.subscribe((params) => {
            this.action = params.get('action');
            if (params.get('id')) {
                this.route.data.subscribe((data) => {
                    if (data.value) {
                        this.form.patchValue(data.value);
                    }
                });
            }
            else {
                this.listRoute = "../list";
            }
            if (this.action === "view") {
                this.form.disable();
            }
            else if (this.form.disabled) {
                this.form.enable();
            }
        });
    }
}
CrudFormComponent.ɵfac = function CrudFormComponent_Factory(t) { return new (t || CrudFormComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc1.ActivatedRoute)); };
CrudFormComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: CrudFormComponent, selectors: [["nCrudForm"]], inputs: { disableSave: "disableSave", form: "form", service: "service" }, ngContentSelectors: _c1, decls: 3, vars: 9, consts: [[3, "form", "action", "route", "disableSave"], ["nButton", "", "pButton", "", "type", "button", "label", "Voltar", 1, "p-mr-2", "p-mb-2", 3, "routerLink"]], template: function CrudFormComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵprojectionDef();
        ɵngcc0.ɵɵelementStart(0, "nForm", 0);
        ɵngcc0.ɵɵprojection(1);
        ɵngcc0.ɵɵelement(2, "button", 1);
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        ɵngcc0.ɵɵproperty("form", ctx.form)("action", ctx.action)("route", ɵngcc0.ɵɵpureFunction1(5, _c0, ctx.listRoute))("disableSave", ctx.disableSave);
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("routerLink", ɵngcc0.ɵɵpureFunction1(7, _c0, ctx.listRoute));
    } }, directives: [ɵngcc2.FormComponent, ɵngcc3.ButtonDirective, ɵngcc1.RouterLink], encapsulation: 2 });
CrudFormComponent.ctorParameters = () => [
    { type: ActivatedRoute }
];
CrudFormComponent.propDecorators = {
    form: [{ type: Input }],
    service: [{ type: Input }],
    disableSave: [{ type: Input }]
};
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(CrudFormComponent, [{
        type: Component,
        args: [{
                selector: 'nCrudForm',
                template: "<nForm [form]=\"form\" [action]=\"action\" [route]=\"[listRoute]\" [disableSave]=\"disableSave\">\n    <ng-content></ng-content>\n    <button nButton pButton type=\"button\" [routerLink]=\"[listRoute]\" label=\"Voltar\" class=\"p-mr-2 p-mb-2\"></button>\n</nForm>\n"
            }]
    }], function () { return [{ type: ɵngcc1.ActivatedRoute }]; }, { disableSave: [{
            type: Input
        }], form: [{
            type: Input
        }], service: [{
            type: Input
        }] }); })();
