// @nuvem/primeng-components/esm2015/lib/crud/form.component.js
import { Component, Inject, Input, Output, EventEmitter } from "@angular/core";
import { CRUD_SERVICE } from "./crud-service.service";
import { Router, ActivatedRoute } from "@angular/router";
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '@angular/router';
import * as ɵngcc2 from '@angular/forms';
import * as ɵngcc3 from '@angular/common';
import * as ɵngcc4 from 'primeng/button';

function FormComponent_div_4_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div");
    ɵngcc0.ɵɵelement(1, "button", 4);
    ɵngcc0.ɵɵprojection(2, 1);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r0 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("label", ctx_r0.saveLabel)("disabled", ctx_r0.form.disabled || ctx_r0.form.invalid);
} }
const _c0 = ["*", [["", "nButton", ""]]];
const _c1 = ["*", "[nButton]"];
export class FormComponent {
    constructor(crudService, activatedRoute, router) {
        this.crudService = crudService;
        this.activatedRoute = activatedRoute;
        this.router = router;
        this.action = "save";
        this.saveLabel = "Salvar";
        this.disableSave = false;
        this.save = new EventEmitter();
        this.reset = new EventEmitter();
    }
    ngOnInit() {
        if (this.service) {
            this.crudService = this.service;
        }
    }
    submit() {
        if (this.save.observers.length) {
            this.save.emit(this.form.value);
        }
        else {
            this.crudService[this.action](this.form.value)
                .subscribe(() => {
                if (this.route) {
                    this.router.navigate(this.route, { relativeTo: this.activatedRoute });
                }
                else if (this.reset.observers.length) {
                    this.reset.emit(this.form);
                }
                else {
                    this.form.reset();
                }
            });
        }
    }
}
FormComponent.ɵfac = function FormComponent_Factory(t) { return new (t || FormComponent)(ɵngcc0.ɵɵdirectiveInject(CRUD_SERVICE), ɵngcc0.ɵɵdirectiveInject(ɵngcc1.ActivatedRoute), ɵngcc0.ɵɵdirectiveInject(ɵngcc1.Router)); };
FormComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: FormComponent, selectors: [["nForm"]], inputs: { action: "action", saveLabel: "saveLabel", disableSave: "disableSave", form: "form", route: "route", service: "service" }, outputs: { save: "save", reset: "reset" }, ngContentSelectors: _c1, decls: 5, vars: 1, consts: [[1, "p-grid"], [1, "p-col-12"], [1, "form-group", 3, "submit"], [4, "ngIf"], ["pButton", "", "type", "submit", 1, "p-mr-2", "p-mb-2", 3, "label", "disabled"]], template: function FormComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵprojectionDef(_c0);
        ɵngcc0.ɵɵelementStart(0, "div", 0);
        ɵngcc0.ɵɵelementStart(1, "div", 1);
        ɵngcc0.ɵɵelementStart(2, "form", 2);
        ɵngcc0.ɵɵlistener("submit", function FormComponent_Template_form_submit_2_listener() { return ctx.submit(); });
        ɵngcc0.ɵɵprojection(3);
        ɵngcc0.ɵɵtemplate(4, FormComponent_div_4_Template, 3, 2, "div", 3);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        ɵngcc0.ɵɵadvance(4);
        ɵngcc0.ɵɵproperty("ngIf", !ctx.disableSave);
    } }, directives: [ɵngcc2.ɵangular_packages_forms_forms_y, ɵngcc2.NgControlStatusGroup, ɵngcc2.NgForm, ɵngcc3.NgIf, ɵngcc4.ButtonDirective], encapsulation: 2 });
FormComponent.ctorParameters = () => [
    { type: undefined, decorators: [{ type: Inject, args: [CRUD_SERVICE,] }] },
    { type: ActivatedRoute },
    { type: Router }
];
FormComponent.propDecorators = {
    form: [{ type: Input }],
    route: [{ type: Input }],
    action: [{ type: Input }],
    saveLabel: [{ type: Input }],
    service: [{ type: Input }],
    disableSave: [{ type: Input }],
    save: [{ type: Output }],
    reset: [{ type: Output }]
};
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(FormComponent, [{
        type: Component,
        args: [{
                selector: 'nForm',
                template: "<div class=\"p-grid\">\n    <div class=\"p-col-12\">\n        <form (submit)=\"submit()\" class=\"form-group\">\n            <ng-content></ng-content>\n            <div *ngIf=\"!disableSave\">\n                <button pButton type=\"submit\" [label]=\"saveLabel\" [disabled]=\"form.disabled || form.invalid\" class=\"p-mr-2 p-mb-2\"></button>\n                <ng-content select=\"[nButton]\"></ng-content>\n            </div>\n        </form>\n    </div>\n</div>"
            }]
    }], function () { return [{ type: undefined, decorators: [{
                type: Inject,
                args: [CRUD_SERVICE]
            }] }, { type: ɵngcc1.ActivatedRoute }, { type: ɵngcc1.Router }]; }, { action: [{
            type: Input
        }], saveLabel: [{
            type: Input
        }], disableSave: [{
            type: Input
        }], save: [{
            type: Output
        }], reset: [{
            type: Output
        }], form: [{
            type: Input
        }], route: [{
            type: Input
        }], service: [{
            type: Input
        }] }); })();
