// @nuvem/angular-base/esm2015/lib/clipboard/clipboard.module.js
import { NgModule } from '@angular/core';
import { ClipboardDirective } from './clipboard.directive';
import { ClipboardService } from './clipboard.service';
/**
 * ClipboardModule expõe o servico ClipboardService para o uso de outras aplicações
 * @module
 */
import * as ɵngcc0 from '@angular/core';
export class ClipboardModule {
}
ClipboardModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: ClipboardModule });
ClipboardModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function ClipboardModule_Factory(t) { return new (t || ClipboardModule)(); }, providers: [
        ClipboardService
    ], imports: [[]] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(ClipboardModule, { declarations: function () { return [ClipboardDirective]; }, exports: function () { return [ClipboardDirective]; } }); })();
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(ClipboardModule, [{
        type: NgModule,
        args: [{
                imports: [],
                declarations: [
                    ClipboardDirective,
                ],
                providers: [
                    ClipboardService
                ],
                exports: [
                    ClipboardDirective,
                ]
            }]
    }], null, null); })();
