// @nuvem/angular-base/esm2015/lib/clipboard/clipboard.directive.js
import { Directive, Input, Output, EventEmitter, HostListener } from '@angular/core';
import { ClipboardService } from './clipboard.service';
/**
 * Diretiva que expõe seletor para o uso do clipbord
 * @class
 */
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from './clipboard.service';
export class ClipboardDirective {
    /**
     * constructor method
     * @param {ClipboardService} service
     * @constructor
     */
    constructor(service) {
        this.service = service;
        /**
         * onClipboard variable
         * @type {EventEmitter<boolean>}
         */
        this.onClipboard = new EventEmitter();
    }
    /**
     * copy method
     * @event click
     * @returns void
     */
    copy() {
        try {
            this.service.copy(this.clipboard);
            this.onClipboard.emit(true);
        }
        catch (e) {
            this.onClipboard.emit(false);
            throw e;
        }
    }
}
ClipboardDirective.ɵfac = function ClipboardDirective_Factory(t) { return new (t || ClipboardDirective)(ɵngcc0.ɵɵdirectiveInject(ɵngcc1.ClipboardService)); };
ClipboardDirective.ɵdir = ɵngcc0.ɵɵdefineDirective({ type: ClipboardDirective, selectors: [["", "clipboard", ""]], hostBindings: function ClipboardDirective_HostBindings(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵlistener("click", function ClipboardDirective_click_HostBindingHandler() { return ctx.copy(); });
    } }, inputs: { clipboard: "clipboard" }, outputs: { onClipboard: "onClipboard" } });
ClipboardDirective.ctorParameters = () => [
    { type: ClipboardService }
];
ClipboardDirective.propDecorators = {
    clipboard: [{ type: Input }],
    onClipboard: [{ type: Output }],
    copy: [{ type: HostListener, args: ['click',] }]
};
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(ClipboardDirective, [{
        type: Directive,
        args: [{
                selector: '[clipboard]'
            }]
    }], function () { return [{ type: ɵngcc1.ClipboardService }]; }, { onClipboard: [{
            type: Output
        }], 
    /**
     * copy method
     * @event click
     * @returns void
     */
    copy: [{
            type: HostListener,
            args: ['click']
        }], clipboard: [{
            type: Input
        }] }); })();
