// @nuvem/angular-base/esm2015/lib/form-utils/injectable-superclass.js
import { Injectable } from "@angular/core";
import { Subject, Subscription } from "rxjs";
/**
 * Use as the superclass for anything managed by angular's dependency injection for care-free use of `subscribeTo()`. It simply calls `unsubscribe()` during `ngOnDestroy()`. If you override `ngOnDestroy()` in your subclass, be sure to invoke the super implementation.
 *
 * ```ts
 * @Injectable()
 * // or @Component() (also consider DirectiveSuperclass)
 * // or @Directive() (also consider DirectiveSuperclass)
 * // or @Pipe()
 * class MyThing extends InjectableSuperclass {
 *   constructor(somethingObservable: Observable) {
 *     super();
 *     this.subscribeTo(somethingObservable);
 *   }
 *
 *   ngOnDestroy() {
 *     // if you override ngOnDestroy, be sure to call this too
 *     super.ngOnDestroy();
 *   }
 * }
 * ```
 */
import * as ɵngcc0 from '@angular/core';
export class InjectableSuperclass {
    constructor() {
        this.subscriptions = new Subscription();
        this.destructionSubject = new Subject();
        this.destruction$ = this.destructionSubject.asObservable();
    }
    subscribeTo(observable, next, error, complete) {
        this.subscriptions.add(observable.subscribe(this.bind(next), this.bind(error), this.bind(complete)));
    }
    unsubscribe() {
        this.subscriptions.unsubscribe();
        this.subscriptions = new Subscription();
    }
    bind(fn) {
        return fn === null || fn === void 0 ? void 0 : fn.bind(this);
    }
    ngOnDestroy() {
        this.unsubscribe();
        this.destructionSubject.next();
        this.destructionSubject.complete();
    }
}
InjectableSuperclass.ɵfac = function InjectableSuperclass_Factory(t) { return new (t || InjectableSuperclass)(); };
InjectableSuperclass.ɵprov = ɵngcc0.ɵɵdefineInjectable({ token: InjectableSuperclass, factory: InjectableSuperclass.ɵfac });
InjectableSuperclass.ctorParameters = () => [];
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(InjectableSuperclass, [{
        type: Injectable
    }], function () { return []; }, null); })();
