// @nuvem/angular-base/esm2015/lib/form-utils/wrapped-form-control-superclass.js
import { Injector, Directive } from "@angular/core";
import { FormControl } from "@angular/forms";
import { FormControlSuperclass } from "./form-control-superclass";
import * as ɵngcc0 from '@angular/core';
export class WrappedFormControlSuperclass extends FormControlSuperclass {
    constructor(injector) {
        super(injector);
        this.formControl = new FormControl();
        this.subscribeTo(this.formControl.valueChanges, (value) => {
            this.emitOutgoingValue(this.innerToOuter(value));
        });
        this.formControl.markAsTouched = () => {
            this.onTouched();
        };
    }
    /** Called as angular propagates values changes to this `ControlValueAccessor`. You normally do not need to use it. */
    handleIncomingValue(value) {
        this.formControl.setValue(this.outerToInner(value), { emitEvent: false });
    }
    /** Called as angular propagates disabled changes to this `ControlValueAccessor`. You normally do not need to use it. */
    setDisabledState(isDisabled) {
        if (isDisabled) {
            this.formControl.disable({ emitEvent: false });
        }
        else {
            this.formControl.enable({ emitEvent: false });
        }
        super.setDisabledState(this.isDisabled);
    }
    /** Override this to modify a value coming from the outside to the format needed within this component. */
    outerToInner(value) {
        return value;
    }
    /** Override this to modify a value coming from within this component to the format expected on the outside. */
    innerToOuter(value) {
        return value;
    }
}
WrappedFormControlSuperclass.ɵfac = function WrappedFormControlSuperclass_Factory(t) { return new (t || WrappedFormControlSuperclass)(ɵngcc0.ɵɵdirectiveInject(ɵngcc0.Injector)); };
WrappedFormControlSuperclass.ɵdir = ɵngcc0.ɵɵdefineDirective({ type: WrappedFormControlSuperclass, features: [ɵngcc0.ɵɵInheritDefinitionFeature] });
WrappedFormControlSuperclass.ctorParameters = () => [
    { type: Injector }
];
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(WrappedFormControlSuperclass, [{
        type: Directive
    }], function () { return [{ type: ɵngcc0.Injector }]; }, null); })();
