// @nuvem/angular-base/esm2015/lib/security/authentication/login-success.component.js
import { Component } from '@angular/core';
import { Authentication } from './authentication';
/**
 * Componente de login que será chamado quando a aplicação for logada com sucesso.
 * Deverá ser criada uma rota para esse componente.
 * @class
 */
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from './authentication';
export class LoginSuccessComponent {
    constructor(authenticationService) {
        this.authenticationService = authenticationService;
    }
    ngOnInit() {
        this.authenticationService.login();
    }
}
LoginSuccessComponent.ɵfac = function LoginSuccessComponent_Factory(t) { return new (t || LoginSuccessComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc1.Authentication)); };
LoginSuccessComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: LoginSuccessComponent, selectors: [["app-login-success"]], decls: 0, vars: 0, template: function LoginSuccessComponent_Template(rf, ctx) { }, encapsulation: 2 });
LoginSuccessComponent.ctorParameters = () => [
    { type: Authentication }
];
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(LoginSuccessComponent, [{
        type: Component,
        args: [{
                selector: 'app-login-success',
                template: ''
            }]
    }], function () { return [{ type: ɵngcc1.Authentication }]; }, null); })();
