// @nuvem/angular-base/esm2015/lib/security/authentication/logout.directive.js
import { Directive, HostListener } from "@angular/core";
import { Authentication } from "./authentication";
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from './authentication';
export class LogoutDirective {
    constructor(authenticationService) {
        this.authenticationService = authenticationService;
    }
    click() {
        this.authenticationService.logout();
    }
}
LogoutDirective.ɵfac = function LogoutDirective_Factory(t) { return new (t || LogoutDirective)(ɵngcc0.ɵɵdirectiveInject(ɵngcc1.Authentication)); };
LogoutDirective.ɵdir = ɵngcc0.ɵɵdefineDirective({ type: LogoutDirective, selectors: [["", "logout", ""]], hostBindings: function LogoutDirective_HostBindings(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵlistener("click", function LogoutDirective_click_HostBindingHandler() { return ctx.click(); });
    } } });
LogoutDirective.ctorParameters = () => [
    { type: Authentication }
];
LogoutDirective.propDecorators = {
    click: [{ type: HostListener, args: ['click',] }]
};
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(LogoutDirective, [{
        type: Directive,
        args: [{ selector: '[logout]' }]
    }], function () { return [{ type: ɵngcc1.Authentication }]; }, { click: [{
            type: HostListener,
            args: ['click']
        }] }); })();
