// @nuvem/angular-base/esm2015/lib/security/authentication/user.directive.js
import { Directive, ElementRef, Input } from "@angular/core";
import { Authentication } from "./authentication";
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from './authentication';
export class UserDirective {
    constructor(el, authenticationService) {
        this.el = el;
        this.authenticationService = authenticationService;
    }
    ngOnInit() {
        if (this.authenticationService.isAuthenticated()) {
            let user = this.authenticationService.getUser();
            this.el.nativeElement.innerHTML += this.user ? user[this.user] : user.name;
        }
    }
}
UserDirective.ɵfac = function UserDirective_Factory(t) { return new (t || UserDirective)(ɵngcc0.ɵɵdirectiveInject(ɵngcc0.ElementRef), ɵngcc0.ɵɵdirectiveInject(ɵngcc1.Authentication)); };
UserDirective.ɵdir = ɵngcc0.ɵɵdefineDirective({ type: UserDirective, selectors: [["", "user", ""]], inputs: { user: "user" } });
UserDirective.ctorParameters = () => [
    { type: ElementRef },
    { type: Authentication }
];
UserDirective.propDecorators = {
    user: [{ type: Input }]
};
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(UserDirective, [{
        type: Directive,
        args: [{ selector: '[user]' }]
    }], function () { return [{ type: ɵngcc0.ElementRef }, { type: ɵngcc1.Authentication }]; }, { user: [{
            type: Input
        }] }); })();
