// @nuvem/angular-base/esm2015/lib/version-tag/version-tag.component.js
import { Component, Input, Inject } from '@angular/core';
import { VERSIONTAG_CONFIG } from './version-tag.config';
/**
 * Componente responsável por expor o seletor <app-version-tag></app-version-tag> para reuso
 * @class
 */
import * as ɵngcc0 from '@angular/core';
export class VersionTagComponent {
    constructor(config) {
        this.config = config;
    }
    /**
     * Metodo executado ao carregar o componente responsável por carregar o arquivo 'package.json'
     * @return void
     */
    ngOnInit() {
        this.version = this.config.appVersion;
    }
}
VersionTagComponent.ɵfac = function VersionTagComponent_Factory(t) { return new (t || VersionTagComponent)(ɵngcc0.ɵɵdirectiveInject(VERSIONTAG_CONFIG)); };
VersionTagComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: VersionTagComponent, selectors: [["app-version-tag"]], inputs: { cssClass: "cssClass" }, decls: 2, vars: 1, consts: [[2, "margin-left", "5px"]], template: function VersionTagComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelementStart(0, "span", 0);
        ɵngcc0.ɵɵtext(1);
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵtextInterpolate1(" Vers\u00E3o: ", ctx.version, "\n");
    } }, encapsulation: 2 });
VersionTagComponent.ctorParameters = () => [
    { type: undefined, decorators: [{ type: Inject, args: [VERSIONTAG_CONFIG,] }] }
];
VersionTagComponent.propDecorators = {
    cssClass: [{ type: Input }]
};
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(VersionTagComponent, [{
        type: Component,
        args: [{
                selector: 'app-version-tag',
                template: "<span style=\"margin-left: 5px;\">\n    Vers\u00E3o: {{version}}\n</span>"
            }]
    }], function () { return [{ type: undefined, decorators: [{
                type: Inject,
                args: [VERSIONTAG_CONFIG]
            }] }]; }, { cssClass: [{
            type: Input
        }] }); })();
