// @nuvem/primeng-components/esm2015/lib/datatable/datatable.component.js
import { Component, ViewChild, Input, Output, ContentChildren, ChangeDetectorRef, EventEmitter, ViewEncapsulation } from '@angular/core';
import { Table as DataTable } from 'primeng/table';
import { DataProviderFactory, DataProviderType } from './dataprovider-factory';
import { DatatableButtonComponent } from './datatable-button.component';
import { DatatableClickEvent } from './datatable-click-event';
import { Authorization } from '@nuvem/angular-base';
import { Column } from './p-column.directive';
/**
 * Class DatatableComponent
 * @class
 */
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from './dataprovider-factory';
import * as ɵngcc2 from '@nuvem/angular-base';
import * as ɵngcc3 from 'primeng/blockui';
import * as ɵngcc4 from 'primeng/progressspinner';
import * as ɵngcc5 from '@angular/common';
import * as ɵngcc6 from 'primeng/button';
import * as ɵngcc7 from 'primeng/table';
import * as ɵngcc8 from 'primeng/api';
import * as ɵngcc9 from './show-column/show-column.component';
import * as ɵngcc10 from './filter-column/filter-column.component';
import * as ɵngcc11 from '@angular/forms';
import * as ɵngcc12 from 'primeng/inputtext';
import * as ɵngcc13 from './action-button/action-button.component';
import * as ɵngcc14 from './datatable-button.component';
import * as ɵngcc15 from '../extra-button/datatable-splitbutton.component';

function DatatableComponent_basis_show_column_9_Template(rf, ctx) { if (rf & 1) {
    const _r10 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "basis-show-column", 18);
    ɵngcc0.ɵɵlistener("updateVisibleColumns", function DatatableComponent_basis_show_column_9_Template_basis_show_column_updateVisibleColumns_0_listener($event) { ɵngcc0.ɵɵrestoreView(_r10); const ctx_r9 = ɵngcc0.ɵɵnextContext(); return ctx_r9.updateVisibleColumns($event); });
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r0 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵproperty("columns", ctx_r0.columns);
} }
function DatatableComponent_ng_template_16_div_0_div_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 25);
    ɵngcc0.ɵɵelementStart(1, "span", 26);
    ɵngcc0.ɵɵtext(2);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r15 = ɵngcc0.ɵɵnextContext(3);
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵtextInterpolate(ctx_r15.horizontalButtonsHeader);
} }
function DatatableComponent_ng_template_16_div_0_ng_container_3_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainer(0);
} }
const _c0 = function (a2, a3) { return { "horizontal-buttons": true, "p-sm-12": true, "p-g-8": a2, "p-g-12": a3 }; };
function DatatableComponent_ng_template_16_div_0_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 22);
    ɵngcc0.ɵɵtemplate(1, DatatableComponent_ng_template_16_div_0_div_1_Template, 3, 1, "div", 23);
    ɵngcc0.ɵɵelementStart(2, "div", 2);
    ɵngcc0.ɵɵtemplate(3, DatatableComponent_ng_template_16_div_0_ng_container_3_Template, 1, 0, "ng-container", 24);
    ɵngcc0.ɵɵprojection(4);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r12 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r12.showHorizontalButtonsHeader);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngClass", ɵngcc0.ɵɵpureFunction2(3, _c0, ctx_r12.showHorizontalButtonsHeader, !ctx_r12.showHorizontalButtonsHeader));
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngTemplateOutlet", ctx_r12.defaultButtons);
} }
function DatatableComponent_ng_template_16_th_2_p_sortIcon_2_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "p-sortIcon", 29);
} if (rf & 2) {
    const col_r17 = ɵngcc0.ɵɵnextContext().$implicit;
    ɵngcc0.ɵɵproperty("field", col_r17.field);
} }
const _c1 = function (a0) { return { "display": a0 }; };
function DatatableComponent_ng_template_16_th_2_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "th", 27);
    ɵngcc0.ɵɵtext(1);
    ɵngcc0.ɵɵtemplate(2, DatatableComponent_ng_template_16_th_2_p_sortIcon_2_Template, 1, 1, "p-sortIcon", 28);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const col_r17 = ctx.$implicit;
    const ctx_r13 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵproperty("pSortableColumn", col_r17.sortable ? col_r17.field : null)("ngStyle", ɵngcc0.ɵɵpureFunction1(4, _c1, ctx_r13.visibleColumns[col_r17.field]));
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵtextInterpolate1(" ", col_r17.header, " ");
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", col_r17.sortable);
} }
function DatatableComponent_ng_template_16_tr_3_th_1_basis_filter_column_1_Template(rf, ctx) { if (rf & 1) {
    const _r25 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "basis-filter-column", 34);
    ɵngcc0.ɵɵlistener("outputEvent", function DatatableComponent_ng_template_16_tr_3_th_1_basis_filter_column_1_Template_basis_filter_column_outputEvent_0_listener($event) { ɵngcc0.ɵɵrestoreView(_r25); const col_r21 = ɵngcc0.ɵɵnextContext().$implicit; const ctx_r23 = ɵngcc0.ɵɵnextContext(3); return ctx_r23.filterColumn(col_r21.field, $event); });
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const col_r21 = ɵngcc0.ɵɵnextContext().$implicit;
    const ctx_r22 = ɵngcc0.ɵɵnextContext(3);
    ɵngcc0.ɵɵproperty("listaOriginal", ctx_r22.filterSelectOptions)("lista", ctx_r22.filterSelectOptions)("field", col_r21.field)("updateFiltro", ctx_r22.updateFilterColumn)("filterType", col_r21.filterType)("tableType", ctx_r22.type)("customFilterOptions", ctx_r22.customFilterOptions);
} }
function DatatableComponent_ng_template_16_tr_3_th_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "th", 32);
    ɵngcc0.ɵɵtemplate(1, DatatableComponent_ng_template_16_tr_3_th_1_basis_filter_column_1_Template, 1, 7, "basis-filter-column", 33);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const col_r21 = ctx.$implicit;
    const ctx_r20 = ɵngcc0.ɵɵnextContext(3);
    ɵngcc0.ɵɵproperty("ngStyle", ɵngcc0.ɵɵpureFunction1(2, _c1, ctx_r20.visibleColumns[col_r21.field]));
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", col_r21.field != "actions");
} }
function DatatableComponent_ng_template_16_tr_3_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "tr", 30);
    ɵngcc0.ɵɵtemplate(1, DatatableComponent_ng_template_16_tr_3_th_1_Template, 2, 4, "th", 31);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const columns_r11 = ɵngcc0.ɵɵnextContext().$implicit;
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngForOf", columns_r11);
} }
function DatatableComponent_ng_template_16_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵtemplate(0, DatatableComponent_ng_template_16_div_0_Template, 5, 6, "div", 19);
    ɵngcc0.ɵɵelementStart(1, "tr");
    ɵngcc0.ɵɵtemplate(2, DatatableComponent_ng_template_16_th_2_Template, 3, 6, "th", 20);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵtemplate(3, DatatableComponent_ng_template_16_tr_3_Template, 2, 1, "tr", 21);
} if (rf & 2) {
    const columns_r11 = ctx.$implicit;
    const ctx_r2 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵproperty("ngIf", ctx_r2.defaultButtons && !ctx_r2.verticalButtons);
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("ngForOf", columns_r11);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r2.filterOnColumn);
} }
function DatatableComponent_ng_template_17_ng_container_1_td_1_ng_container_1_span_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "span", 41);
} if (rf & 2) {
    const col_r33 = ɵngcc0.ɵɵnextContext(3).$implicit;
    const rowData_r28 = ɵngcc0.ɵɵnextContext().$implicit;
    ɵngcc0.ɵɵproperty("innerHTML", rowData_r28[col_r33.field], ɵngcc0.ɵɵsanitizeHtml);
} }
const _c2 = function (a0, a1) { return { rowData: a0, col: a1 }; };
function DatatableComponent_ng_template_17_ng_container_1_td_1_ng_container_1_ng_container_2_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainer(0, 42);
} if (rf & 2) {
    const col_r33 = ɵngcc0.ɵɵnextContext(3).$implicit;
    const rowData_r28 = ɵngcc0.ɵɵnextContext().$implicit;
    ɵngcc0.ɵɵproperty("ngTemplateOutlet", col_r33.bodyTemplate)("ngTemplateOutletContext", ɵngcc0.ɵɵpureFunction2(2, _c2, rowData_r28, col_r33));
} }
function DatatableComponent_ng_template_17_ng_container_1_td_1_ng_container_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵtemplate(1, DatatableComponent_ng_template_17_ng_container_1_td_1_ng_container_1_span_1_Template, 1, 1, "span", 39);
    ɵngcc0.ɵɵtemplate(2, DatatableComponent_ng_template_17_ng_container_1_td_1_ng_container_1_ng_container_2_Template, 1, 5, "ng-container", 40);
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const col_r33 = ɵngcc0.ɵɵnextContext(2).$implicit;
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", !col_r33.bodyTemplate);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", col_r33.bodyTemplate);
} }
function DatatableComponent_ng_template_17_ng_container_1_td_1_ng_container_2_ng_template_2_span_0_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "span", 41);
} if (rf & 2) {
    const col_r33 = ɵngcc0.ɵɵnextContext(4).$implicit;
    const rowData_r28 = ɵngcc0.ɵɵnextContext().$implicit;
    ɵngcc0.ɵɵproperty("innerHTML", rowData_r28[col_r33.field], ɵngcc0.ɵɵsanitizeHtml);
} }
const _c3 = function () { return { "width": "100%" }; };
function DatatableComponent_ng_template_17_ng_container_1_td_1_ng_container_2_ng_template_2_div_1_Template(rf, ctx) { if (rf & 1) {
    const _r55 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "div", 46);
    ɵngcc0.ɵɵelementStart(1, "div", 47);
    ɵngcc0.ɵɵelementStart(2, "div", 46);
    ɵngcc0.ɵɵelementStart(3, "span", 48);
    ɵngcc0.ɵɵelementStart(4, "input", 49);
    ɵngcc0.ɵɵlistener("ngModelChange", function DatatableComponent_ng_template_17_ng_container_1_td_1_ng_container_2_ng_template_2_div_1_Template_input_ngModelChange_4_listener($event) { ɵngcc0.ɵɵrestoreView(_r55); const col_r33 = ɵngcc0.ɵɵnextContext(4).$implicit; const rowData_r28 = ɵngcc0.ɵɵnextContext().$implicit; return (rowData_r28[col_r33.field] = $event); });
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(5, "label");
    ɵngcc0.ɵɵtext(6);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const col_r33 = ɵngcc0.ɵɵnextContext(4).$implicit;
    const rowData_r28 = ɵngcc0.ɵɵnextContext().$implicit;
    ɵngcc0.ɵɵadvance(4);
    ɵngcc0.ɵɵstyleMap(ɵngcc0.ɵɵpureFunction0(4, _c3));
    ɵngcc0.ɵɵproperty("ngModel", rowData_r28[col_r33.field]);
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵtextInterpolate(col_r33.header);
} }
const _c4 = function (a0, a1) { return { "rowData": a0, "col": a1 }; };
function DatatableComponent_ng_template_17_ng_container_1_td_1_ng_container_2_ng_template_2_ng_container_2_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainer(0, 42);
} if (rf & 2) {
    const col_r33 = ɵngcc0.ɵɵnextContext(4).$implicit;
    const rowData_r28 = ɵngcc0.ɵɵnextContext().$implicit;
    ɵngcc0.ɵɵproperty("ngTemplateOutlet", col_r33.inputTemplate)("ngTemplateOutletContext", ɵngcc0.ɵɵpureFunction2(2, _c4, rowData_r28, col_r33));
} }
function DatatableComponent_ng_template_17_ng_container_1_td_1_ng_container_2_ng_template_2_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵtemplate(0, DatatableComponent_ng_template_17_ng_container_1_td_1_ng_container_2_ng_template_2_span_0_Template, 1, 1, "span", 39);
    ɵngcc0.ɵɵtemplate(1, DatatableComponent_ng_template_17_ng_container_1_td_1_ng_container_2_ng_template_2_div_1_Template, 7, 5, "div", 45);
    ɵngcc0.ɵɵtemplate(2, DatatableComponent_ng_template_17_ng_container_1_td_1_ng_container_2_ng_template_2_ng_container_2_Template, 1, 5, "ng-container", 40);
} if (rf & 2) {
    const col_r33 = ɵngcc0.ɵɵnextContext(3).$implicit;
    ɵngcc0.ɵɵproperty("ngIf", !col_r33.outputTemplate && col_r33.editable == false);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", !col_r33.inputTemplate && col_r33.editable != false);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", col_r33.inputTemplate);
} }
function DatatableComponent_ng_template_17_ng_container_1_td_1_ng_container_2_ng_template_3_span_0_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "span", 41);
} if (rf & 2) {
    const col_r33 = ɵngcc0.ɵɵnextContext(4).$implicit;
    const rowData_r28 = ɵngcc0.ɵɵnextContext().$implicit;
    ɵngcc0.ɵɵproperty("innerHTML", rowData_r28[col_r33.field], ɵngcc0.ɵɵsanitizeHtml);
} }
function DatatableComponent_ng_template_17_ng_container_1_td_1_ng_container_2_ng_template_3_ng_container_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainer(0, 42);
} if (rf & 2) {
    const col_r33 = ɵngcc0.ɵɵnextContext(4).$implicit;
    const rowData_r28 = ɵngcc0.ɵɵnextContext().$implicit;
    ɵngcc0.ɵɵproperty("ngTemplateOutlet", col_r33.outputTemplate)("ngTemplateOutletContext", ɵngcc0.ɵɵpureFunction2(2, _c2, rowData_r28, col_r33));
} }
function DatatableComponent_ng_template_17_ng_container_1_td_1_ng_container_2_ng_template_3_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵtemplate(0, DatatableComponent_ng_template_17_ng_container_1_td_1_ng_container_2_ng_template_3_span_0_Template, 1, 1, "span", 39);
    ɵngcc0.ɵɵtemplate(1, DatatableComponent_ng_template_17_ng_container_1_td_1_ng_container_2_ng_template_3_ng_container_1_Template, 1, 5, "ng-container", 40);
} if (rf & 2) {
    const col_r33 = ɵngcc0.ɵɵnextContext(3).$implicit;
    ɵngcc0.ɵɵproperty("ngIf", !col_r33.outputTemplate);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", col_r33.outputTemplate);
} }
function DatatableComponent_ng_template_17_ng_container_1_td_1_ng_container_2_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵelementStart(1, "p-cellEditor");
    ɵngcc0.ɵɵtemplate(2, DatatableComponent_ng_template_17_ng_container_1_td_1_ng_container_2_ng_template_2_Template, 3, 3, "ng-template", 43);
    ɵngcc0.ɵɵtemplate(3, DatatableComponent_ng_template_17_ng_container_1_td_1_ng_container_2_ng_template_3_Template, 2, 2, "ng-template", 44);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementContainerEnd();
} }
function DatatableComponent_ng_template_17_ng_container_1_td_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "td", 32);
    ɵngcc0.ɵɵtemplate(1, DatatableComponent_ng_template_17_ng_container_1_td_1_ng_container_1_Template, 3, 2, "ng-container", 15);
    ɵngcc0.ɵɵtemplate(2, DatatableComponent_ng_template_17_ng_container_1_td_1_ng_container_2_Template, 4, 0, "ng-container", 15);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const col_r33 = ɵngcc0.ɵɵnextContext().$implicit;
    const ctx_r35 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵproperty("ngStyle", ɵngcc0.ɵɵpureFunction1(3, _c1, ctx_r35.visibleColumns[col_r33.field]));
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", !ctx_r35.editMode);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r35.editMode);
} }
const _c5 = function (a0) { return { rowData: a0 }; };
function DatatableComponent_ng_template_17_ng_container_1_td_2_ng_container_2_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainer(0, 42);
} if (rf & 2) {
    const col_r33 = ɵngcc0.ɵɵnextContext(2).$implicit;
    const rowData_r28 = ɵngcc0.ɵɵnextContext().$implicit;
    ɵngcc0.ɵɵproperty("ngTemplateOutlet", col_r33.outputTemplate)("ngTemplateOutletContext", ɵngcc0.ɵɵpureFunction1(2, _c5, rowData_r28));
} }
function DatatableComponent_ng_template_17_ng_container_1_td_2_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "td", 50);
    ɵngcc0.ɵɵelementStart(1, "basis-action-button");
    ɵngcc0.ɵɵtemplate(2, DatatableComponent_ng_template_17_ng_container_1_td_2_ng_container_2_Template, 1, 4, "ng-container", 40);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const col_r33 = ɵngcc0.ɵɵnextContext().$implicit;
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("ngIf", col_r33.outputTemplate);
} }
function DatatableComponent_ng_template_17_ng_container_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵtemplate(1, DatatableComponent_ng_template_17_ng_container_1_td_1_Template, 3, 5, "td", 37);
    ɵngcc0.ɵɵtemplate(2, DatatableComponent_ng_template_17_ng_container_1_td_2_Template, 3, 1, "td", 38);
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const col_r33 = ctx.$implicit;
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", col_r33.field != "actions");
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", col_r33.field == "actions");
} }
const _c6 = function (a0) { return { "editing": a0 }; };
function DatatableComponent_ng_template_17_Template(rf, ctx) { if (rf & 1) {
    const _r75 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "tr", 35);
    ɵngcc0.ɵɵlistener("mouseover", function DatatableComponent_ng_template_17_Template_tr_mouseover_0_listener($event) { ɵngcc0.ɵɵrestoreView(_r75); const ctx_r74 = ɵngcc0.ɵɵnextContext(); return ctx_r74.editarLinhaOver($event); })("mouseleave", function DatatableComponent_ng_template_17_Template_tr_mouseleave_0_listener($event) { ɵngcc0.ɵɵrestoreView(_r75); const ctx_r76 = ɵngcc0.ɵɵnextContext(); return ctx_r76.editarLinhaOut($event); });
    ɵngcc0.ɵɵtemplate(1, DatatableComponent_ng_template_17_ng_container_1_Template, 3, 2, "ng-container", 36);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const rowData_r28 = ctx.$implicit;
    const columns_r29 = ctx.columns;
    const editing_r30 = ctx.editing;
    ɵngcc0.ɵɵproperty("pSelectableRow", rowData_r28)("pContextMenuRow", rowData_r28)("pEditableRow", rowData_r28)("ngClass", ɵngcc0.ɵɵpureFunction1(5, _c6, editing_r30));
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngForOf", columns_r29);
} }
function DatatableComponent_18_ng_template_0_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵtext(0);
} if (rf & 2) {
    const ctx_r77 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵtextInterpolate1(" ", ctx_r77.emptyMessage, " ");
} }
function DatatableComponent_18_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵtemplate(0, DatatableComponent_18_ng_template_0_Template, 1, 1, "ng-template");
} }
function DatatableComponent_p_footer_19_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "p-footer");
    ɵngcc0.ɵɵtext(1);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r5 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵtextInterpolate1(" ", ctx_r5.paginationFooterText(ctx_r5.footerMsg), " ");
} }
function DatatableComponent_div_20_ng_container_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainer(0);
} }
function DatatableComponent_div_20_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 51);
    ɵngcc0.ɵɵtemplate(1, DatatableComponent_div_20_ng_container_1_Template, 1, 0, "ng-container", 24);
    ɵngcc0.ɵɵprojection(2, 1);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r6 = ɵngcc0.ɵɵnextContext();
    const _r7 = ɵngcc0.ɵɵreference(22);
    ɵngcc0.ɵɵstyleProp("position", ctx_r6.enableButtonsScroll ? "sticky" : "")("top", ctx_r6.enableButtonsScroll ? "15%" : "");
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngTemplateOutlet", _r7);
} }
function DatatableComponent_ng_template_21_basis_datatable_button_0_Template(rf, ctx) { if (rf & 1) {
    const _r84 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "basis-datatable-button", 56);
    ɵngcc0.ɵɵlistener("click", function DatatableComponent_ng_template_21_basis_datatable_button_0_Template_basis_datatable_button_click_0_listener($event) { ɵngcc0.ɵɵrestoreView(_r84); const ctx_r83 = ɵngcc0.ɵɵnextContext(2); return ctx_r83.onClick("edit", $event); });
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r79 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵclassProp("horizontal-button", !ctx_r79.verticalButtons);
    ɵngcc0.ɵɵproperty("bTooltip", ctx_r79.selectedRow != null ? "Editar" : "Escolha um registro para editar.")("disabled", ctx_r79.selectedRow == null);
} }
function DatatableComponent_ng_template_21_basis_datatable_button_1_Template(rf, ctx) { if (rf & 1) {
    const _r86 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "basis-datatable-button", 57);
    ɵngcc0.ɵɵlistener("click", function DatatableComponent_ng_template_21_basis_datatable_button_1_Template_basis_datatable_button_click_0_listener($event) { ɵngcc0.ɵɵrestoreView(_r86); const ctx_r85 = ɵngcc0.ɵɵnextContext(2); return ctx_r85.onClick("view", $event); });
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r80 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵclassProp("horizontal-button", !ctx_r80.verticalButtons);
    ɵngcc0.ɵɵproperty("bTooltip", ctx_r80.selectedRow != null ? "Visualizar" : "Escolha um registro para visualizar.")("disabled", ctx_r80.selectedRow == null);
} }
function DatatableComponent_ng_template_21_basis_datatable_button_2_Template(rf, ctx) { if (rf & 1) {
    const _r88 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "basis-datatable-button", 58);
    ɵngcc0.ɵɵlistener("click", function DatatableComponent_ng_template_21_basis_datatable_button_2_Template_basis_datatable_button_click_0_listener($event) { ɵngcc0.ɵɵrestoreView(_r88); const ctx_r87 = ɵngcc0.ɵɵnextContext(2); return ctx_r87.onClick("delete", $event); });
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r81 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵclassProp("horizontal-button", !ctx_r81.verticalButtons);
    ɵngcc0.ɵɵproperty("bTooltip", ctx_r81.selectedRow != null ? "Excluir" : "Escolha um registro para excluir.")("disabled", ctx_r81.selectedRow == null);
} }
function DatatableComponent_ng_template_21_datatable_splitbutton_3_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "datatable-splitbutton", 59);
} if (rf & 2) {
    const ctx_r82 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵclassMap(ctx_r82.splitButton.class);
    ɵngcc0.ɵɵproperty("icon", ctx_r82.splitButton.icon)("items", ctx_r82.splitButton.itens)("label", ctx_r82.splitButton.label)("method", ctx_r82.splitButton.method);
} }
function DatatableComponent_ng_template_21_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵtemplate(0, DatatableComponent_ng_template_21_basis_datatable_button_0_Template, 1, 4, "basis-datatable-button", 52);
    ɵngcc0.ɵɵtemplate(1, DatatableComponent_ng_template_21_basis_datatable_button_1_Template, 1, 4, "basis-datatable-button", 53);
    ɵngcc0.ɵɵtemplate(2, DatatableComponent_ng_template_21_basis_datatable_button_2_Template, 1, 4, "basis-datatable-button", 54);
    ɵngcc0.ɵɵtemplate(3, DatatableComponent_ng_template_21_datatable_splitbutton_3_Template, 1, 6, "datatable-splitbutton", 55);
} if (rf & 2) {
    const ctx_r8 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵproperty("ngIf", !ctx_r8.disableEdit && ctx_r8.checkRolesToButton("rulesToShowView") && ctx_r8.checkRolesToButton("rulesToHideView"));
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", !ctx_r8.disableView && ctx_r8.checkRolesToButton("rulesToShowEdit") && ctx_r8.checkRolesToButton("rulesToHideEdit"));
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", !ctx_r8.disableDelete && ctx_r8.checkRolesToButton("rulesToShowDelete") && ctx_r8.checkRolesToButton("rulesToHideDelete"));
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r8.splitButton);
} }
const _c7 = [[["horizontalButtons"]], [["basis-datatable-button"]]];
const _c8 = function (a0) { return { "active": a0 }; };
const _c9 = function (a0, a1, a2, a3) { return { "p-col-11": a0, "p-col-12": a1, "edicao-tabela": a2, "editing-row": a3 }; };
const _c10 = ["horizontalButtons", "basis-datatable-button"];
export class DatatableComponent {
    /**
     * constructor method
     * @constructor
     * @param {ChangeDetectorRef} changeDetectorRef
     * @param {DataProviderFactory} dataProviderFactory
     * @param {Authorization} authorizationService
     */
    constructor(changeDetectorRef, dataProviderFactory, authorizationService) {
        this.changeDetectorRef = changeDetectorRef;
        this.dataProviderFactory = dataProviderFactory;
        this.authorizationService = authorizationService;
        this.isShowGlobalFilter = false;
        this.isShowFilterRow = false;
        /**
         * filterParams property
         * @type {any}
         */
        this.filterParams = {};
        /**
         * footerMsg property
         * @type {string}
         */
        this.footerMsg = '';
        /**
         * orderInSort property
         * @type {boolean}
         */
        this.orderInSort = false;
        /**
         * value property
         * @type {any}
         */
        this.value = [];
        /**
         * type property
         * @type DataProviderType
         */
        this.type = DataProviderType.Server;
        /**
         * rows property
         * @type {number}
         */
        this.rows = 5;
        /**
         * showPaginationFooter property
         * @type {boolean}
         */
        this.showPaginationFooter = false;
        /**
         * disableEdit property
         * @type {boolean}
         */
        this.disableEdit = false;
        /**
         * disableView property
         * @type {boolean}
         */
        this.disableView = false;
        /**
         * disableDelete property
         * @type {boolean}
         */
        this.disableDelete = false;
        /**
         * selectionMode property
         * @type {string}
         */
        this.selectionMode = 'single';
        /**
         * enableButtonsScroll property
         * @type {boolean}
         */
        this.enableButtonsScroll = false;
        /**
         * emptyMessage property
         * @type {string}
         */
        this.emptyMessage = 'Nenhum registro encontrado.';
        /**
         * verticalButtons property
         * @type {boolean}
         */
        this.verticalButtons = true;
        /**
         * defaultButtons property
         * @type {boolean}
         */
        this.defaultButtons = true;
        /**
         * enableScroll property
         * @type {boolean}
         */
        this.enableScroll = false;
        /**
         * disableLoadingBlockUI property
         * @type {boolean}
         */
        this.disableLoadingBlockUI = false;
        /**
         * filterOnColumn property
         */
        this.filterOnColumn = false;
        /**
         * filterSelectOptions property
         */
        this.filterSelectOptions = [];
        /**
         * updateFiltroHistorico property
         */
        this.updateFilterColumn = false;
        /**
         * customFilterOptions property
         */
        this.customFilterOptions = {};
        /**
         * buttonClick property
         * @type {EventEmitter<any>}
         */
        this.buttonClick = new EventEmitter();
        /**
         * startedLoading property
         * @type {EventEmitter<void>}
         */
        this.startedLoading = new EventEmitter();
        /**
         * finishedLoading property
         * @type {EventEmitter<HttpErrorResponse>}
         */
        this.finishedLoading = new EventEmitter();
        /**
         * onRowSelected property
         * @type {EventEmitter<any>}
         */
        this.onRowSelected = new EventEmitter();
        /**
         * showHorizontalButtonsHeader property
         * @type {boolean}
         */
        this.showHorizontalButtonsHeader = false;
        /**
         * blockContent
         * @type {boolean}
         */
        this.blockContent = false;
        /**
         * visibleColumns property
         * @type {Object}
         */
        this.visibleColumns = {};
        /**
         * showVisibleColumnsControl property
         * @type {boolean}
         */
        this.showVisibleColumnsControl = false;
        /**
         * dataKey property
         * Identificador único para cada registro da tabela
         * @type {string}
         */
        this.dataKey = 'id';
        /**
         * onRowEditEvent property
         * Evento disparado quando o usuário realiza ações de edição em uma linha da tabela
         * @type {EventEmitter<void>}
         */
        this.onRowEditEvent = new EventEmitter();
        this.clonedRowData = {};
        /**
         * errorMessage property
         * @type {string}
         */
        this.errorMessage = 'Ocorreu um erro ao carregar os dados da tabela. Tente novamente.';
        this.editingRow = false;
    }
    get showGlobalFilter() {
        return this.isShowGlobalFilter;
    }
    set showGlobalFilter(showGlobalFilter) {
        this.isShowGlobalFilter = showGlobalFilter;
    }
    get showFilterRow() {
        return this.isShowFilterRow;
    }
    set showFilterRow(ShowFilterRow) {
        this.isShowFilterRow = ShowFilterRow;
    }
    /**
     * ngOnInit method
     * @returns void
     */
    ngOnInit() {
        this.dataProvider = this.dataProviderFactory.create(this.type, this);
        this.pDatatableComponent.rows = this.rows;
        this.subscribeToLoadingEvents();
        this.initRowsPerPageOptionsIfPresent();
        this.defineIfShouldShowHorizontalButtonsHeader();
        this.updateOnValueChange();
        this.finishedLoading.subscribe(() => {
            this.updateOnValueChange();
        });
    }
    /**
     * subscribeFromValueChange method
     * @returns void
     */
    updateOnValueChange() {
        this.filterSelectOptions = [];
        if (this.value) {
            this.value.map((item) => {
                this.filterSelectOptions.push(item);
            });
        }
    }
    /**
     * subscribeToLoadingEvents method
     * @returns void
     */
    subscribeToLoadingEvents() {
        if (!this.disableLoadingBlockUI) {
            this.originalEmptyMessage = this.emptyMessage;
            this.startedLoading.subscribe(() => this.onStartedLoading());
            this.finishedLoading.subscribe(error => this.onFinishedLoading(error));
        }
    }
    /**
     * onStartedLoading method
     * @returns void
     */
    onStartedLoading() {
        this.blockContent = true;
        this.emptyMessage = this.originalEmptyMessage;
    }
    /**
     * onFinishedLoading method
     * @param {HttpErrorResponse} error
     * @returns void
     */
    onFinishedLoading(error) {
        this.blockContent = false;
        if (error) {
            this.emptyMessage = this.errorMessage;
        }
    }
    /**
     * initRowsPerPageOptionsIfPresent method
     * @returns void
     */
    initRowsPerPageOptionsIfPresent() {
        if (this.rowsPerPageOptions) {
            this.pDatatableComponent.rowsPerPageOptions = this.rowsPerPageOptions;
        }
    }
    /**
     * defineIfShouldShowHorizontalButtonsHeader method
     * @returns void
     */
    defineIfShouldShowHorizontalButtonsHeader() {
        if (this.horizontalButtonsHeader) {
            this.showHorizontalButtonsHeader = true;
        }
    }
    /**
     * ngAfterViewInit method
     * @returns void
     */
    ngAfterViewInit() {
        this.setExtraParams();
        this.pDatatableComponent.columns = this.columns;
        this.pDatatableComponent.ngAfterContentInit();
        this.setAllColumnsVisible();
        this.changeDetectorRef.detectChanges();
    }
    /**
     * setExtraParams method
     * @returns void
     */
    setExtraParams() {
        for (let i in this.extraParams) {
            this.pDatatableComponent[i] = this.extraParams[i];
        }
    }
    /**
     * ngAfterContentInit method
     * @returns void
     */
    ngAfterContentInit() {
        this.extraButtons.forEach(button => {
            button.click.subscribe(event => {
                this.onClick(button.name, event);
            });
        });
    }
    /**
     * onClick method
     * @param {string} button
     * @param {any} event
     * @return void
     */
    onClick(button, event) {
        this.buttonClick.emit(new DatatableClickEvent(button, this.selectedRow));
        event.stopPropagation();
        this.resetSelectedRow(button);
    }
    /**
     * onSelectChange method
     * @param {any} event
     * @returns void
     */
    onSelectChange(event) {
        this.onRowSelected.emit(this.selectedRow);
        if (Array.isArray(this.selectedRow) && this.selectedRow.length > 1) {
            this.updateMultipleButtonsStatus(true);
        }
        else {
            this.updateMultipleButtonsStatus(false);
        }
    }
    /**
     * updateMultipleButtonsStatus method
     * @param {boolean} hasMultipleSelection
     * @return void
     */
    updateMultipleButtonsStatus(hasMultipleSelection) {
        this.extraButtons.forEach(button => {
            if (hasMultipleSelection && !button.allowMultipleSelection) {
                button.disabled = true;
            }
            else {
                button.disabled = false;
            }
        });
    }
    /**
     * refresh method
     * @param query
     * @returns void
     */
    refresh(query) {
        this.dataProvider.load(query);
    }
    /**
     * resetSelectedRow method
     * @param button
     * @returns void
     */
    resetSelectedRow(button) {
        if (button === 'delete') {
            this.selectedRow = null;
        }
    }
    /**
     * reset method
     * @returns void
     */
    reset() {
        this.pDatatableComponent.first = 0;
        this.dataProvider.reset();
        this.refresh();
    }
    /**
     * paginationFooterText method
     * @param {string} msg
     * @returns string
     */
    paginationFooterText(msg) {
        const registros = this.totalRecords() === 1 ? 'registro' : 'registros';
        if (msg === '') {
            return `Exibindo: ${this.currentFirstIndex()} a ${this.currentLastIndex()} de ${this.totalRecords()} ${registros}`;
        }
        msg = msg.replace('{totalRecords}', this.totalRecords().toString());
        msg = msg.replace('{currentFirstIndex}', this.currentFirstIndex().toString());
        msg = msg.replace('{currentLastIndex}', this.currentLastIndex().toString());
        return msg;
    }
    /**
     * currentFirstIndex method
     * @returns number
     */
    currentFirstIndex() {
        return this.pDatatableComponent.first + 1;
    }
    /**
     * currentLastIndex method
     * @returns number
     */
    currentLastIndex() {
        const currentLastIndex = this.pDatatableComponent.first + this.pDatatableComponent.rows;
        if (currentLastIndex > this.totalRecords()) {
            return this.totalRecords();
        }
        return currentLastIndex;
    }
    /**
     * totalRecords method
     * @returns number
     */
    totalRecords() {
        return this.pDatatableComponent.totalRecords;
    }
    /**
     * checkRolesToButton method
     * @param {string} button
     * @return boolean
     */
    checkRolesToButton(button) {
        if (this[button]) {
            if (button === 'rulesToHide' ||
                button === 'rulesToShowEdit' ||
                button === 'rulesToShowDelete') {
                this.checkUseButtons(button);
                return this.authorizationService.hasRole(this[button]);
            }
            return !this.authorizationService.hasRole(this[button]);
        }
        return true;
    }
    /**
     * checkUseButtons method
     * @param {string} button
     * @returns void
     */
    checkUseButtons(button) {
        if (this[button] && this[button.replace('Show', 'Hide')]) {
            throw new Error('Não é possível definir dois valores de roles para os botões padrões.');
        }
    }
    /**
     * filter method
     * @returns void
     */
    filter() {
        for (let prop in this.filterParams) {
            if (this.filterParams[prop] === '' ||
                this.filterParams[prop].length === 0) {
                delete this.filterParams[prop];
            }
        }
        if (Object.getOwnPropertyNames(this.filterParams).length === 0) {
            this.reset();
        }
        else {
            this.refresh(this.filterParams);
        }
    }
    /**
     * Método para filtrar por colunas
     * @param column chave da coluna na tabela
     * @param filters valor do filtro
     */
    filterColumn(column, filters) {
        const values = [];
        if (filters instanceof Array) {
            values[column] = [];
            for (let item in filters) {
                let val = filters[item] instanceof Object
                    ? filters[item].value
                    : filters[item];
                values[column].push(val);
            }
        }
        else {
            values[column] = filters;
        }
        this.filterParams = Object.assign(Object.assign({}, this.filterParams), values);
        return this.filter();
    }
    /**
     * Inicia todas as colunas da tabela como visíveis
     */
    setAllColumnsVisible() {
        this.columns.forEach(col => {
            this.visibleColumns[col.field] = 'table-cell';
        });
    }
    /**
     * Verifica se uma coluna deve estar visível
     * @param column string
     * @param visibleColumns Array
     */
    visibleColumnCheck(column, visibleColumns) {
        return visibleColumns.some((item) => {
            return item.field ? item.field == column : true;
        });
    }
    /**
     * Atualzia os valores da variável 'visibleColuns' para mostrar ou não as colunas
     * @param columns array
     */
    updateVisibleColumns(columns) {
        for (let col in this.visibleColumns) {
            if (this.visibleColumnCheck(col, columns)) {
                this.visibleColumns[col] = 'table-cell';
            }
            else {
                this.visibleColumns[col] = 'none';
            }
        }
    }
    onRowEditInitEvent(event, rowData) {
        // const tr = event.srcElement.closest('tr');
        // setTimeout(() => {
        //     this.editingRow = true;
        //     const top = tr.offsetTop;
        //     const height = tr.offsetHeight;
        //     const editButtons = tr.querySelector('.editar-campos');
        //     editButtons.style.top = (top + height) + 'px';
        //     editButtons.style.height ='75px';
        //     editButtons.style.backgroundColor = '#d0d0d0';
        //     editButtons.style.opacity = 1;
        //     this.clonedRowData[rowData.id] = {...rowData};
        //     this.onRowEditEvent.emit({event: "init", data: rowData});
        // }, null , tr);
    }
    onRowEditSaveEvent(event, rowData) {
        this.onRowEditEvent.emit({ event: "save", data: rowData });
        this.resetEditingMode(event);
    }
    onRowEditCancelEvent(event, rowData, index) {
        this.value[index] = this.clonedRowData[rowData.id];
        delete this.clonedRowData[rowData[this.dataKey]];
        this.onRowEditEvent.emit({ event: "cancel", data: rowData });
        this.resetEditingMode(event);
    }
    resetEditingMode(event) {
        const editButtons = event.srcElement.closest('tr').querySelector('.editar-campos');
        editButtons.style.opacity = 0;
        editButtons.style.backgroundColor = '#0000006b';
        editButtons.style.display = 'none';
        this.editingRow = false;
    }
    editarLinhaOver(event) {
        // if (!this.editingRow && this.editMode == 'row') {
        //     const top = event.srcElement.closest('tr').offsetTop;
        //     const height = event.srcElement.closest('tr').offsetHeight;
        //     const editButtons = event.srcElement.closest('tr').querySelector('.editar-campos');
        //     editButtons.style.display = 'flex';
        //     editButtons.style.top = top + 'px';
        //     editButtons.style.height = height + 'px';
        //     editButtons.style.backgroundColor = '#0000006b';
        //     editButtons.style.opacity = 1;
        // }
    }
    editarLinhaOut(event) {
        // if (!this.editingRow) {
        //     const editButtons = event.srcElement.closest('tr').querySelector('.editar-campos');
        //     editButtons.style.display = 'none';
        //     editButtons.style.opacity = 0;
        // }
    }
}
DatatableComponent.ɵfac = function DatatableComponent_Factory(t) { return new (t || DatatableComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc0.ChangeDetectorRef), ɵngcc0.ɵɵdirectiveInject(ɵngcc1.DataProviderFactory), ɵngcc0.ɵɵdirectiveInject(ɵngcc2.Authorization)); };
DatatableComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: DatatableComponent, selectors: [["basis-datatable"]], contentQueries: function DatatableComponent_ContentQueries(rf, ctx, dirIndex) { if (rf & 1) {
        ɵngcc0.ɵɵcontentQuery(dirIndex, DatatableButtonComponent, 0);
        ɵngcc0.ɵɵcontentQuery(dirIndex, Column, 0);
    } if (rf & 2) {
        let _t;
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.extraButtons = _t);
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.columns = _t);
    } }, viewQuery: function DatatableComponent_Query(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵviewQuery(DataTable, 3);
    } if (rf & 2) {
        let _t;
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.pDatatableComponent = _t.first);
    } }, inputs: { value: "value", type: "type", rows: "rows", showPaginationFooter: "showPaginationFooter", disableEdit: "disableEdit", disableView: "disableView", disableDelete: "disableDelete", selectionMode: "selectionMode", enableButtonsScroll: "enableButtonsScroll", emptyMessage: "emptyMessage", verticalButtons: "verticalButtons", defaultButtons: "defaultButtons", enableScroll: "enableScroll", disableLoadingBlockUI: "disableLoadingBlockUI", filterOnColumn: "filterOnColumn", customFilterOptions: "customFilterOptions", showVisibleColumnsControl: "showVisibleColumnsControl", dataKey: "dataKey", title: "title", splitButton: "splitButton", rowsPerPageOptions: "rowsPerPageOptions", url: "url", extraParams: "extraParams", paginationParameters: "paginationParameters", scrollHeight: "scrollHeight", scrollWidth: "scrollWidth", rulesToHide: "rulesToHide", rulesToShowEdit: "rulesToShowEdit", rulesToShowDelete: "rulesToShowDelete", rulesToHideView: "rulesToHideView", rulesToHideEdit: "rulesToHideEdit", rulesToHideDelete: "rulesToHideDelete", rowStyleClass: "rowStyleClass", editMode: "editMode" }, outputs: { buttonClick: "buttonClick", startedLoading: "startedLoading", finishedLoading: "finishedLoading", onRowSelected: "onRowSelected", onRowEditEvent: "onRowEditEvent" }, ngContentSelectors: _c10, decls: 23, vars: 29, consts: [[3, "blocked"], [1, "center-spinner"], [3, "ngClass"], [1, "table-tools"], [1, "p-col-8", "nopad"], [1, ""], [1, "p-col-4", "nopad", "tools"], [3, "columns", "updateVisibleColumns", 4, "ngIf"], ["pButton", "", "pRipple", "", "type", "button", "icon", "fas fa-sliders-h", 1, "p-button-rounded", "p-button-outlined", "p-mr-2", "p-mb-2", 3, "ngClass", "click"], ["pButton", "", "pRipple", "", "type", "button", "icon", "fas fa-columns", 1, "p-button-rounded", "p-button-outlined", "p-mr-2", "p-mb-2", 3, "ngClass", "click"], [1, "p-grid"], ["styleClass", "p-datatable-customers", 3, "value", "rows", "paginator", "selectionMode", "responsive", "scrollable", "selection", "editMode", "rowHover", "dataKey", "selectionChange", "onRowSelect", "onRowUnselect"], ["datatable", ""], ["pTemplate", "header"], ["pTemplate", "body"], [4, "ngIf"], ["class", "p-col-1", 3, "position", "top", 4, "ngIf"], ["defaultButtonsTemplate", ""], [3, "columns", "updateVisibleColumns"], ["class", "p-g", 4, "ngIf"], [3, "pSortableColumn", "ngStyle", 4, "ngFor", "ngForOf"], ["class", "header-filter-column", 4, "ngIf"], [1, "p-g"], ["class", "p-g-4 p-sm-12", 4, "ngIf"], [4, "ngTemplateOutlet"], [1, "p-g-4", "p-sm-12"], [1, "horizontal-buttons-header"], [3, "pSortableColumn", "ngStyle"], [3, "field", 4, "ngIf"], [3, "field"], [1, "header-filter-column"], [3, "ngStyle", 4, "ngFor", "ngForOf"], [3, "ngStyle"], [3, "listaOriginal", "lista", "field", "updateFiltro", "filterType", "tableType", "customFilterOptions", "outputEvent", 4, "ngIf"], [3, "listaOriginal", "lista", "field", "updateFiltro", "filterType", "tableType", "customFilterOptions", "outputEvent"], [3, "pSelectableRow", "pContextMenuRow", "pEditableRow", "ngClass", "mouseover", "mouseleave"], [4, "ngFor", "ngForOf"], [3, "ngStyle", 4, "ngIf"], ["width", "8%", "class", "action-column", 4, "ngIf"], [3, "innerHTML", 4, "ngIf"], [3, "ngTemplateOutlet", "ngTemplateOutletContext", 4, "ngIf"], [3, "innerHTML"], [3, "ngTemplateOutlet", "ngTemplateOutletContext"], ["pTemplate", "input"], ["pTemplate", "output"], ["class", "p-md-12", 4, "ngIf"], [1, "p-md-12"], [1, "p-g", "form-group"], [1, "md-inputfield"], ["pInputText", "", "type", "text", 3, "ngModel", "ngModelChange"], ["width", "8%", 1, "action-column"], [1, "p-col-1"], ["class", "p-button-success", "icon", "pi pi-pencil", 3, "bTooltip", "horizontal-button", "disabled", "click", 4, "ngIf"], ["icon", "pi pi-search-plus", 3, "bTooltip", "horizontal-button", "disabled", "click", 4, "ngIf"], ["class", "p-button-danger", "icon", "pi pi-trash", 3, "bTooltip", "horizontal-button", "disabled", "click", 4, "ngIf"], [3, "icon", "class", "items", "label", "method", 4, "ngIf"], ["icon", "pi pi-pencil", 1, "p-button-success", 3, "bTooltip", "disabled", "click"], ["icon", "pi pi-search-plus", 3, "bTooltip", "disabled", "click"], ["icon", "pi pi-trash", 1, "p-button-danger", 3, "bTooltip", "disabled", "click"], [3, "icon", "items", "label", "method"]], template: function DatatableComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵprojectionDef(_c7);
        ɵngcc0.ɵɵelementStart(0, "p-blockUI", 0);
        ɵngcc0.ɵɵelementStart(1, "div", 1);
        ɵngcc0.ɵɵelement(2, "p-progressSpinner");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(3, "div", 2);
        ɵngcc0.ɵɵelementStart(4, "div", 3);
        ɵngcc0.ɵɵelementStart(5, "div", 4);
        ɵngcc0.ɵɵelementStart(6, "h3", 5);
        ɵngcc0.ɵɵtext(7);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(8, "div", 6);
        ɵngcc0.ɵɵtemplate(9, DatatableComponent_basis_show_column_9_Template, 1, 1, "basis-show-column", 7);
        ɵngcc0.ɵɵelementStart(10, "button", 8);
        ɵngcc0.ɵɵlistener("click", function DatatableComponent_Template_button_click_10_listener() { return ctx.filterOnColumn = !ctx.filterOnColumn; });
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(11, "button", 9);
        ɵngcc0.ɵɵlistener("click", function DatatableComponent_Template_button_click_11_listener() { return ctx.showVisibleColumnsControl = !ctx.showVisibleColumnsControl; });
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(12, "div", 10);
        ɵngcc0.ɵɵelementStart(13, "div", 2);
        ɵngcc0.ɵɵelementStart(14, "p-table", 11, 12);
        ɵngcc0.ɵɵlistener("selectionChange", function DatatableComponent_Template_p_table_selectionChange_14_listener($event) { return ctx.selectedRow = $event; })("onRowSelect", function DatatableComponent_Template_p_table_onRowSelect_14_listener($event) { return ctx.onSelectChange($event); })("onRowUnselect", function DatatableComponent_Template_p_table_onRowUnselect_14_listener($event) { return ctx.onSelectChange($event); });
        ɵngcc0.ɵɵtemplate(16, DatatableComponent_ng_template_16_Template, 4, 3, "ng-template", 13);
        ɵngcc0.ɵɵtemplate(17, DatatableComponent_ng_template_17_Template, 2, 7, "ng-template", 14);
        ɵngcc0.ɵɵtemplate(18, DatatableComponent_18_Template, 1, 0, undefined, 15);
        ɵngcc0.ɵɵtemplate(19, DatatableComponent_p_footer_19_Template, 2, 1, "p-footer", 15);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵtemplate(20, DatatableComponent_div_20_Template, 3, 5, "div", 16);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵtemplate(21, DatatableComponent_ng_template_21_Template, 4, 4, "ng-template", null, 17, ɵngcc0.ɵɵtemplateRefExtractor);
    } if (rf & 2) {
        ɵngcc0.ɵɵproperty("blocked", ctx.blockContent);
        ɵngcc0.ɵɵadvance(3);
        ɵngcc0.ɵɵproperty("ngClass", ctx.defaultButtons && ctx.verticalButtons ? "p-grid p-col-11" : "p-grid p-col-12");
        ɵngcc0.ɵɵadvance(4);
        ɵngcc0.ɵɵtextInterpolate(ctx.title);
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("ngIf", ctx.showVisibleColumnsControl);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngClass", ɵngcc0.ɵɵpureFunction1(20, _c8, ctx.filterOnColumn));
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngClass", ɵngcc0.ɵɵpureFunction1(22, _c8, ctx.showVisibleColumnsControl));
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("ngClass", ɵngcc0.ɵɵpureFunction4(24, _c9, ctx.defaultButtons && ctx.verticalButtons, !ctx.defaultButtons || !ctx.verticalButtons, ctx.editMode == "row", ctx.editingRow));
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("value", ctx.value)("rows", ctx.rows)("paginator", true)("selectionMode", ctx.selectionMode)("responsive", true)("scrollable", ctx.enableScroll)("selection", ctx.selectedRow)("editMode", ctx.editMode)("rowHover", true)("dataKey", ctx.dataKey);
        ɵngcc0.ɵɵadvance(4);
        ɵngcc0.ɵɵproperty("ngIf", ctx.emptyMessage);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", ctx.showPaginationFooter && ctx.totalRecords() > 0);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", ctx.defaultButtons && ctx.verticalButtons);
    } }, directives: [ɵngcc3.BlockUI, ɵngcc4.ProgressSpinner, ɵngcc5.NgClass, ɵngcc5.NgIf, ɵngcc6.ButtonDirective, ɵngcc7.Table, ɵngcc8.PrimeTemplate, ɵngcc9.ShowColumn, ɵngcc5.NgForOf, ɵngcc5.NgTemplateOutlet, ɵngcc7.SortableColumn, ɵngcc5.NgStyle, ɵngcc7.SortIcon, ɵngcc10.FilterColumn, ɵngcc7.SelectableRow, ɵngcc7.ContextMenuRow, ɵngcc7.EditableRow, ɵngcc7.CellEditor, ɵngcc11.DefaultValueAccessor, ɵngcc12.InputText, ɵngcc11.NgControlStatus, ɵngcc11.NgModel, ɵngcc13.ActionButtonComponent, ɵngcc8.Footer, ɵngcc14.DatatableButtonComponent, ɵngcc15.DatatableSplitbuttonComponent], styles: ["body .p-datatable .p-datatable-data tr td{word-wrap:break-word}.horizontal-buttons{display:flex;flex-flow:row wrap;justify-content:flex-end}.horizontal-button{padding:0 3px}.horizontal-buttons-header{font-size:1.3em;word-wrap:break-word}.center-spinner{display:flex;flex-direction:column;justify-content:center;align-items:center;text-align:center;min-height:100vh}.text-right{text-align:right}basis-show-column .p-multiselect-panel.p-widget{width:100%;max-width:100%}basis-show-column{padding-right:15px;border-right:1px solid #747474;margin-right:10px}.table-tools{width:100%;text-align:right;margin-bottom:-10px;display:flex;align-items:center}.table-tools h3{text-align:left;float:left;margin:0}.table-tools a{color:#16695c;border-radius:100px;width:20px;height:20px;padding:6px;display:inline-table;margin-left:3px;font-size:15px;text-align:center;line-height:20px}.table-tools a:hover{color:#1aa590;cursor:pointer}.table-tools a.active{color:#16695c;background-color:#e8e8e8;box-shadow:inset 0 2px 3px rgba(0,0,0,.4392156862745098)}.table-tools .tools{display:inline-flex;align-items:center;justify-content:flex-end}.edicao-tabela,.edita-linha{position:relative}tr.editing td{padding-bottom:75px!important}.editar-campos{opacity:0;position:absolute;width:100%;background-color:rgba(0,0,0,.4196078431372549);left:0;margin-top:0;display:flex;align-items:center;justify-content:center;min-height:35px;transition:all .3s ease-in-out;z-index:1}.editar-campos button{color:#fff;font-size:14px;padding:6px 30px 10px;border-radius:5px;height:28px;width:120px!important}.bt-cancelar,.bt-salvar{margin:0 10px}"], encapsulation: 2 });
DatatableComponent.ctorParameters = () => [
    { type: ChangeDetectorRef },
    { type: DataProviderFactory },
    { type: Authorization }
];
DatatableComponent.propDecorators = {
    title: [{ type: Input }],
    splitButton: [{ type: Input }],
    value: [{ type: Input }],
    type: [{ type: Input }],
    rows: [{ type: Input }],
    rowsPerPageOptions: [{ type: Input }],
    url: [{ type: Input }],
    extraParams: [{ type: Input }],
    paginationParameters: [{ type: Input }],
    showPaginationFooter: [{ type: Input }],
    disableEdit: [{ type: Input }],
    disableView: [{ type: Input }],
    disableDelete: [{ type: Input }],
    selectionMode: [{ type: Input }],
    enableButtonsScroll: [{ type: Input }],
    emptyMessage: [{ type: Input }],
    verticalButtons: [{ type: Input }],
    defaultButtons: [{ type: Input }],
    enableScroll: [{ type: Input }],
    scrollHeight: [{ type: Input }],
    scrollWidth: [{ type: Input }],
    rulesToHide: [{ type: Input }],
    rulesToShowEdit: [{ type: Input }],
    rulesToShowDelete: [{ type: Input }],
    rulesToHideView: [{ type: Input }],
    rulesToHideEdit: [{ type: Input }],
    rulesToHideDelete: [{ type: Input }],
    disableLoadingBlockUI: [{ type: Input }],
    rowStyleClass: [{ type: Input }],
    filterOnColumn: [{ type: Input }],
    customFilterOptions: [{ type: Input }],
    buttonClick: [{ type: Output }],
    pDatatableComponent: [{ type: ViewChild, args: [DataTable, { static: true },] }],
    extraButtons: [{ type: ContentChildren, args: [DatatableButtonComponent,] }],
    columns: [{ type: ContentChildren, args: [Column,] }],
    startedLoading: [{ type: Output }],
    finishedLoading: [{ type: Output }],
    onRowSelected: [{ type: Output }],
    showVisibleColumnsControl: [{ type: Input }],
    editMode: [{ type: Input }],
    dataKey: [{ type: Input }],
    onRowEditEvent: [{ type: Output }]
};
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(DatatableComponent, [{
        type: Component,
        args: [{
                selector: 'basis-datatable',
                template: "<p-blockUI [blocked]=\"blockContent\">\n    <div class=\"center-spinner\">\n        <p-progressSpinner></p-progressSpinner>\n    </div>\n</p-blockUI>\n<div [ngClass]=\"(defaultButtons && verticalButtons) ? 'p-grid p-col-11' : 'p-grid p-col-12'\">\n    <div class=\"table-tools\">\n        <div class=\"p-col-8 nopad\">\n            <h3 class=\"\">{{ title }}</h3>\n        </div>\n        <div class=\"p-col-4 nopad tools\">\n            <basis-show-column *ngIf=\"showVisibleColumnsControl\" [columns]=\"columns\" (updateVisibleColumns)=\"updateVisibleColumns($event)\"></basis-show-column>\n            <button pButton \n                pRipple \n                type=\"button\" \n                icon=\"fas fa-sliders-h\"\n                class=\"p-button-rounded p-button-outlined p-mr-2 p-mb-2\" \n                [ngClass]=\"{'active':filterOnColumn}\"\n                (click)=\"filterOnColumn = !this.filterOnColumn\">\n            </button>\n            <button pButton \n                pRipple \n                type=\"button\" \n                icon=\"fas fa-columns\" \n                [ngClass]=\"{'active':showVisibleColumnsControl}\"\n                class=\"p-button-rounded p-button-outlined p-mr-2 p-mb-2\"\n                (click)=\"showVisibleColumnsControl = !this.showVisibleColumnsControl\">\n            </button>\n        </div>\n    </div>\n</div>\n<div class=\"p-grid\">\n    <div [ngClass]=\"{\n            'p-col-11' : (defaultButtons && verticalButtons),\n            'p-col-12' : (!defaultButtons || !verticalButtons),\n            'edicao-tabela' : editMode == 'row',\n            'editing-row': editingRow\n        }\">\n        <p-table\n            [value]=\"value\"\n            [rows]=\"rows\"\n            [paginator]=\"true\"\n            [selectionMode]=\"selectionMode\"\n            [responsive]=\"true\"\n            [scrollable]=\"enableScroll\"\n            [(selection)]=\"selectedRow\"\n            [editMode]=\"editMode\"\n            (onRowSelect)=\"onSelectChange($event)\"\n            (onRowUnselect)=\"onSelectChange($event)\"\n            styleClass=\"p-datatable-customers\" \n            [rowHover]=\"true\"\n            #datatable\n            [dataKey]=\"dataKey\">\n            <ng-template pTemplate=\"header\" let-columns>\n                <div class=\"p-g\" *ngIf=\"defaultButtons && !verticalButtons\">\n                    <div *ngIf=\"showHorizontalButtonsHeader\" class=\"p-g-4 p-sm-12\">\n                        <span class=\"horizontal-buttons-header\">{{ horizontalButtonsHeader }}</span>\n                    </div>\n                    <div [ngClass]=\"{\n                        'horizontal-buttons': true,\n                        'p-sm-12': true,\n                        'p-g-8': showHorizontalButtonsHeader,\n                        'p-g-12': !showHorizontalButtonsHeader}\">\n                        <ng-container *ngTemplateOutlet=\"defaultButtons\"></ng-container>\n                        <ng-content select=\"horizontalButtons\"></ng-content>\n                    </div>\n                </div>\n                <tr>\n                    <th *ngFor=\"let col of columns\"\n                        [pSortableColumn]=\"col.sortable ? col.field : null\"\n                        [ngStyle]=\"{'display': visibleColumns[col.field]}\">\n                        {{col.header}}\n                        <p-sortIcon *ngIf=\"col.sortable\" [field]=\"col.field\"></p-sortIcon>\n                    </th>\n                </tr>\n                <tr class=\"header-filter-column\" *ngIf=\"filterOnColumn\">\n                    <th *ngFor=\"let col of columns\" [ngStyle]=\"{'display': visibleColumns[col.field]}\">\n                        <basis-filter-column\n                            *ngIf=\"col.field != 'actions'\"\n                            [listaOriginal]=\"filterSelectOptions\"\n                            [lista]=\"filterSelectOptions\"\n                            [field]=\"col.field\"\n                            [updateFiltro]=\"updateFilterColumn\"\n                            (outputEvent)=\"filterColumn(col.field, $event)\"\n                            [filterType]=\"col.filterType\"\n                            [tableType]=\"type\"\n                            [customFilterOptions]=\"customFilterOptions\">\n                        </basis-filter-column>\n                    </th>\n                </tr>\n            </ng-template>\n            <ng-template pTemplate=\"body\" let-rowData let-columns=\"columns\" let-editing=\"editing\" let-ri=\"rowIndex\">\n                <tr [pSelectableRow]=\"rowData\"\n                    [pContextMenuRow]=\"rowData\"\n                    [pEditableRow]=\"rowData\"\n                    [ngClass]=\"{'editing': editing}\"\n                    (mouseover)=\"editarLinhaOver($event)\"\n                    (mouseleave)=\"editarLinhaOut($event)\">\n                    <ng-container *ngFor=\"let col of columns; last as isLast\">\n                        <td *ngIf=\"col.field != 'actions'\" [ngStyle]=\"{'display': visibleColumns[col.field]}\">\n                            <ng-container *ngIf=\"!editMode\">\n                                <span *ngIf=\"!col.bodyTemplate\" [innerHTML]=\"rowData[col.field]\"></span>\n                                <ng-container\n                                    *ngIf=\"col.bodyTemplate\"\n                                    [ngTemplateOutlet]=\"col.bodyTemplate\"\n                                    [ngTemplateOutletContext]=\"{ rowData: rowData, col: col }\">\n                                </ng-container>\n                            </ng-container>\n        \n                            <ng-container *ngIf=\"editMode\">\n                                <p-cellEditor>\n                                    <ng-template pTemplate=\"input\">\n                                        <span *ngIf=\"!col.outputTemplate && col.editable == false\" [innerHTML]=\"rowData[col.field]\"></span>\n                                        <div class=\"p-md-12\" *ngIf=\"!col.inputTemplate && col.editable != false\">\n                                            <div class=\"p-g form-group\">\n                                                <div class=\"p-md-12\">\n                                                    <span class=\"md-inputfield\">\n                                                        <input pInputText type=\"text\" [(ngModel)]=\"rowData[col.field]\" [style]=\"{'width':'100%'}\">\n                                                        <label>{{ col.header }}</label>\n                                                    </span>\n                                                </div>\n                                            </div>\n                                        </div>\n                                        <ng-container\n                                            *ngIf=\"col.inputTemplate\"\n                                            [ngTemplateOutlet]=\"col.inputTemplate\"\n                                            [ngTemplateOutletContext]=\"{ 'rowData': rowData, 'col': col }\">\n                                        </ng-container>\n                                    </ng-template>\n                                    <ng-template pTemplate=\"output\">\n                                        <span *ngIf=\"!col.outputTemplate\" [innerHTML]=\"rowData[col.field]\"></span>\n                                        <ng-container\n                                            *ngIf=\"col.outputTemplate\"\n                                            [ngTemplateOutlet]=\"col.outputTemplate\"\n                                            [ngTemplateOutletContext]=\"{ rowData: rowData, col: col }\">\n                                        </ng-container>\n                                    </ng-template>\n                                </p-cellEditor>\n                            </ng-container>\n                        </td>\n                        <td width=\"8%\" class=\"action-column\" *ngIf=\"col.field == 'actions'\">\n                            <basis-action-button>\n                                <ng-container\n                                    *ngIf=\"col.outputTemplate\"\n                                    [ngTemplateOutlet]=\"col.outputTemplate\"\n                                    [ngTemplateOutletContext]=\"{ rowData: rowData }\">\n                                </ng-container>\n                            </basis-action-button>\n                        </td>\n                        <!-- TODO: verificar implementa\u00E7\u00E3o -->\n                        <!-- <div *ngIf=\"isLast\" class=\"editar-campos\">\n                            <button *ngIf=\"!editing\" pButton type=\"button\" pInitEditableRow (click)=\"onRowEditInitEvent($event, rowData)\">Editar</button>\n                            <button *ngIf=\"editing\" pButton type=\"button\" pSaveEditableRow class=\"p-button-success\" style=\"margin-right: .5em\" (click)=\"onRowEditSaveEvent($event, rowData)\">Salvar</button>\n                            <button *ngIf=\"editing\" pButton type=\"button\" pCancelEditableRow class=\"p-button-danger\" (click)=\"onRowEditCancelEvent($event, rowData, ri)\">Cancelar</button>\n                        </div> -->\n                    </ng-container>\n                </tr>\n            </ng-template>\n            <ng-template *ngIf=\"emptyMessage\">\n                {{ emptyMessage }}\n            </ng-template>\n            <p-footer *ngIf=\"showPaginationFooter && totalRecords() > 0\">\n                {{ paginationFooterText(footerMsg) }}\n            </p-footer>\n        </p-table>\n    </div>\n    <div *ngIf=\"defaultButtons && verticalButtons\"\n        class=\"p-col-1\"\n        [style.position]=\"enableButtonsScroll ? 'sticky' : ''\"\n        [style.top]=\"enableButtonsScroll ? '15%' : ''\">\n            <ng-container *ngTemplateOutlet=\"defaultButtonsTemplate\"></ng-container>\n            <ng-content select=\"basis-datatable-button\"></ng-content>\n    </div>\n</div>\n<ng-template #defaultButtonsTemplate>\n    <basis-datatable-button [bTooltip]=\"selectedRow != null ? 'Editar' : 'Escolha um registro para editar.' \"\n        [class.horizontal-button]=\"!verticalButtons\" class=\"p-button-success\" icon=\"pi pi-pencil\" (click)=\"onClick('edit', $event)\"\n        *ngIf=\"!disableEdit && checkRolesToButton('rulesToShowView') && checkRolesToButton('rulesToHideView')\" [disabled]=\"selectedRow == null\">\n    </basis-datatable-button>\n\n    <basis-datatable-button [bTooltip]=\"selectedRow != null ? 'Visualizar' : 'Escolha um registro para visualizar.'\"\n        [class.horizontal-button]=\"!verticalButtons\" icon=\"pi pi-search-plus\" (click)=\"onClick('view', $event)\"\n        *ngIf=\"!disableView && checkRolesToButton('rulesToShowEdit') && checkRolesToButton('rulesToHideEdit')\" [disabled]=\"selectedRow == null\">\n    </basis-datatable-button>\n\n    <basis-datatable-button [bTooltip]=\"selectedRow != null ? 'Excluir' : 'Escolha um registro para excluir.'\"\n        [class.horizontal-button]=\"!verticalButtons\" class=\"p-button-danger\" icon=\"pi pi-trash\" (click)=\"onClick('delete', $event)\"\n        *ngIf=\"!disableDelete && checkRolesToButton('rulesToShowDelete') && checkRolesToButton('rulesToHideDelete')\" [disabled]=\"selectedRow == null\">\n    </basis-datatable-button>\n\n    <datatable-splitbutton\n        [icon]=\"splitButton.icon\"\n        [class]=\"splitButton.class\"\n        [items]=\"splitButton.itens\"\n        [label]=\"splitButton.label\"\n        [method]=\"splitButton.method\"\n        *ngIf=\"splitButton\">\n    </datatable-splitbutton>\n</ng-template>\n",
                encapsulation: ViewEncapsulation.None,
                styles: ["body .p-datatable .p-datatable-data tr td{word-wrap:break-word}.horizontal-buttons{display:flex;flex-flow:row wrap;justify-content:flex-end}.horizontal-button{padding:0 3px}.horizontal-buttons-header{font-size:1.3em;word-wrap:break-word}.center-spinner{display:flex;flex-direction:column;justify-content:center;align-items:center;text-align:center;min-height:100vh}.text-right{text-align:right}basis-show-column .p-multiselect-panel.p-widget{width:100%;max-width:100%}basis-show-column{padding-right:15px;border-right:1px solid #747474;margin-right:10px}.table-tools{width:100%;text-align:right;margin-bottom:-10px;display:flex;align-items:center}.table-tools h3{text-align:left;float:left;margin:0}.table-tools a{color:#16695c;border-radius:100px;width:20px;height:20px;padding:6px;display:inline-table;margin-left:3px;font-size:15px;text-align:center;line-height:20px}.table-tools a:hover{color:#1aa590;cursor:pointer}.table-tools a.active{color:#16695c;background-color:#e8e8e8;box-shadow:inset 0 2px 3px rgba(0,0,0,.4392156862745098)}.table-tools .tools{display:inline-flex;align-items:center;justify-content:flex-end}.edicao-tabela,.edita-linha{position:relative}tr.editing td{padding-bottom:75px!important}.editar-campos{opacity:0;position:absolute;width:100%;background-color:rgba(0,0,0,.4196078431372549);left:0;margin-top:0;display:flex;align-items:center;justify-content:center;min-height:35px;transition:all .3s ease-in-out;z-index:1}.editar-campos button{color:#fff;font-size:14px;padding:6px 30px 10px;border-radius:5px;height:28px;width:120px!important}.bt-cancelar,.bt-salvar{margin:0 10px}"]
            }]
    }], function () { return [{ type: ɵngcc0.ChangeDetectorRef }, { type: ɵngcc1.DataProviderFactory }, { type: ɵngcc2.Authorization }]; }, { value: [{
            type: Input
        }], type: [{
            type: Input
        }], rows: [{
            type: Input
        }], showPaginationFooter: [{
            type: Input
        }], disableEdit: [{
            type: Input
        }], disableView: [{
            type: Input
        }], disableDelete: [{
            type: Input
        }], selectionMode: [{
            type: Input
        }], enableButtonsScroll: [{
            type: Input
        }], emptyMessage: [{
            type: Input
        }], verticalButtons: [{
            type: Input
        }], defaultButtons: [{
            type: Input
        }], enableScroll: [{
            type: Input
        }], disableLoadingBlockUI: [{
            type: Input
        }], filterOnColumn: [{
            type: Input
        }], customFilterOptions: [{
            type: Input
        }], buttonClick: [{
            type: Output
        }], startedLoading: [{
            type: Output
        }], finishedLoading: [{
            type: Output
        }], onRowSelected: [{
            type: Output
        }], showVisibleColumnsControl: [{
            type: Input
        }], dataKey: [{
            type: Input
        }], onRowEditEvent: [{
            type: Output
        }], title: [{
            type: Input
        }], splitButton: [{
            type: Input
        }], rowsPerPageOptions: [{
            type: Input
        }], url: [{
            type: Input
        }], extraParams: [{
            type: Input
        }], paginationParameters: [{
            type: Input
        }], scrollHeight: [{
            type: Input
        }], scrollWidth: [{
            type: Input
        }], rulesToHide: [{
            type: Input
        }], rulesToShowEdit: [{
            type: Input
        }], rulesToShowDelete: [{
            type: Input
        }], rulesToHideView: [{
            type: Input
        }], rulesToHideEdit: [{
            type: Input
        }], rulesToHideDelete: [{
            type: Input
        }], rowStyleClass: [{
            type: Input
        }], pDatatableComponent: [{
            type: ViewChild,
            args: [DataTable, { static: true }]
        }], extraButtons: [{
            type: ContentChildren,
            args: [DatatableButtonComponent]
        }], columns: [{
            type: ContentChildren,
            args: [Column]
        }], editMode: [{
            type: Input
        }] }); })();
