// @nuvem/primeng-components/esm2015/lib/datatable/datatable-button.component.js
import { Component, Input, Output, EventEmitter } from '@angular/core';
/**
 * Class DatatableButtonComponent
 * @class
 */
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from 'primeng/tooltip';
import * as ɵngcc2 from 'primeng/button';
export class DatatableButtonComponent {
    constructor() {
        /**
         * allowMultipleSelection property
         * @type {boolean}
         */
        this.allowMultipleSelection = true;
        /**
         * disabled property
         * @type boolean
         */
        this.disabled = false;
        /**
         * click property
         * @type {EventEmitter<any>}
         */
        this.click = new EventEmitter();
    }
    /**
     * onClick method
     * @param {any} event
     * @return void
     */
    onClick(event) {
        this.click.emit(event);
        event.stopPropagation();
    }
}
DatatableButtonComponent.ɵfac = function DatatableButtonComponent_Factory(t) { return new (t || DatatableButtonComponent)(); };
DatatableButtonComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: DatatableButtonComponent, selectors: [["basis-datatable-button"]], inputs: { allowMultipleSelection: "allowMultipleSelection", disabled: "disabled", icon: "icon", class: "class", bTooltip: "bTooltip", name: "name" }, outputs: { click: "click" }, decls: 2, vars: 6, consts: [[3, "pTooltip"], ["pRipple", "", "pButton", "", "type", "button", 3, "disabled", "icon", "click"]], template: function DatatableButtonComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelementStart(0, "p", 0);
        ɵngcc0.ɵɵelementStart(1, "button", 1);
        ɵngcc0.ɵɵlistener("click", function DatatableButtonComponent_Template_button_click_1_listener($event) { return ctx.onClick($event); });
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        ɵngcc0.ɵɵpropertyInterpolate("pTooltip", ctx.bTooltip);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵclassMapInterpolate1("p-button-rounded p-mr-2 p-mb-2 ", ctx.class, "");
        ɵngcc0.ɵɵpropertyInterpolate("icon", ctx.icon);
        ɵngcc0.ɵɵproperty("disabled", ctx.disabled);
    } }, directives: [ɵngcc1.Tooltip, ɵngcc2.ButtonDirective], encapsulation: 2 });
DatatableButtonComponent.propDecorators = {
    icon: [{ type: Input }],
    class: [{ type: Input }],
    bTooltip: [{ type: Input }],
    name: [{ type: Input }],
    allowMultipleSelection: [{ type: Input }],
    disabled: [{ type: Input }],
    click: [{ type: Output }]
};
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(DatatableButtonComponent, [{
        type: Component,
        args: [{
                selector: 'basis-datatable-button',
                template: "<p pTooltip=\"{{ bTooltip }}\">\n    <button pRipple\n        (click)=\"onClick($event)\" \n        [disabled]=\"disabled\" \n        pButton \n        type=\"button\" \n        icon=\"{{ icon }}\" \n        class=\"p-button-rounded p-mr-2 p-mb-2 {{ class }}\">\n    </button>\n</p>\n"
            }]
    }], function () { return []; }, { allowMultipleSelection: [{
            type: Input
        }], disabled: [{
            type: Input
        }], click: [{
            type: Output
        }], icon: [{
            type: Input
        }], class: [{
            type: Input
        }], bTooltip: [{
            type: Input
        }], name: [{
            type: Input
        }] }); })();
