// @nuvem/primeng-components/esm2015/lib/datatable/datatable.module.js
import { HttpClientModule, HttpClient } from '@angular/common/http';
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { ButtonModule } from 'primeng/button';
import { TooltipModule } from 'primeng/tooltip';
import { BlockUIModule } from 'primeng/blockui';
import { ProgressSpinnerModule } from 'primeng/progressspinner';
import { SplitButtonModule } from 'primeng/splitbutton';
import { InputTextModule } from 'primeng/inputtext';
import { PaginatorModule } from 'primeng/paginator';
import { TableModule } from 'primeng/table';
import { DataProviderFactory } from './dataprovider-factory';
import { DatatableComponent } from './datatable.component';
import { DatatableButtonComponent } from './datatable-button.component';
import { DatatableService } from './datatable.service';
import { DatatableSplitbuttonModule } from '../extra-button/datatable-splitbutton.module';
import { FilterColumnModule } from './filter-column/filter-column.module';
import { ShowColumnModule } from './show-column/show-column.module';
import { Column } from './p-column.directive';
import { ActionButtonComponent } from './action-button/action-button.component';
/**
 * Class DatatableModule
 * @class
 */
import * as ɵngcc0 from '@angular/core';
export class DatatableModule {
}
DatatableModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: DatatableModule });
DatatableModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function DatatableModule_Factory(t) { return new (t || DatatableModule)(); }, providers: [
        { provide: DatatableService, useClass: DatatableService, deps: [HttpClient] },
        DataProviderFactory
    ], imports: [[
            CommonModule,
            ButtonModule,
            SplitButtonModule,
            TooltipModule,
            TableModule,
            BlockUIModule,
            ProgressSpinnerModule,
            DatatableSplitbuttonModule,
            HttpClientModule,
            PaginatorModule,
            FilterColumnModule,
            ShowColumnModule,
            InputTextModule
        ], FilterColumnModule,
        ShowColumnModule] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(DatatableModule, { declarations: function () { return [DatatableComponent,
        DatatableButtonComponent,
        Column,
        ActionButtonComponent]; }, imports: function () { return [CommonModule,
        ButtonModule,
        SplitButtonModule,
        TooltipModule,
        TableModule,
        BlockUIModule,
        ProgressSpinnerModule,
        DatatableSplitbuttonModule,
        HttpClientModule,
        PaginatorModule,
        FilterColumnModule,
        ShowColumnModule,
        InputTextModule]; }, exports: function () { return [DatatableComponent,
        DatatableButtonComponent,
        FilterColumnModule,
        ShowColumnModule,
        Column]; } }); })();
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(DatatableModule, [{
        type: NgModule,
        args: [{
                imports: [
                    CommonModule,
                    ButtonModule,
                    SplitButtonModule,
                    TooltipModule,
                    TableModule,
                    BlockUIModule,
                    ProgressSpinnerModule,
                    DatatableSplitbuttonModule,
                    HttpClientModule,
                    PaginatorModule,
                    FilterColumnModule,
                    ShowColumnModule,
                    InputTextModule
                ],
                declarations: [
                    DatatableComponent,
                    DatatableButtonComponent,
                    Column,
                    ActionButtonComponent
                ],
                providers: [
                    { provide: DatatableService, useClass: DatatableService, deps: [HttpClient] },
                    DataProviderFactory
                ],
                exports: [
                    DatatableComponent,
                    DatatableButtonComponent,
                    FilterColumnModule,
                    ShowColumnModule,
                    Column
                ]
            }]
    }], null, null); })();
