// @nuvem/primeng-components/esm2015/lib/block-ui/block-ui.component.js
import { BlockUiService } from '@nuvem/angular-base';
import { Component } from '@angular/core';
/**
 * Class BlockUiComponent
 * @class
 */
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '@nuvem/angular-base';
import * as ɵngcc2 from 'primeng/blockui';
import * as ɵngcc3 from 'primeng/progressspinner';
export class BlockUiComponent {
    /**
     * constructor method
     * @constructor
     * @param {BlockUiService} service
     */
    constructor(service) {
        this.service = service;
        /**
         * status property
         * @type {boolean}
         * @public
         */
        this.status = false;
    }
    /**
     * ngOnInit method
     * @returns void
     */
    ngOnInit() {
        this.subscription = this.service.loaderStatus.subscribe((blockui) => {
            this.status = blockui.status;
        });
    }
    /**
     * ngOnDestroy methods
     * @returns void
     */
    ngOnDestroy() {
        this.subscription.unsubscribe();
    }
    /**
     * isBlocked method
     * @returns boolean
     */
    isBlocked() {
        return this.status;
    }
    /**
     * show method
     * @returns false
     */
    show() {
        this.status = true;
    }
    /**
     * hide methods
     * @returns void
     */
    hide() {
        this.status = false;
    }
}
BlockUiComponent.ɵfac = function BlockUiComponent_Factory(t) { return new (t || BlockUiComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc1.BlockUiService)); };
BlockUiComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: BlockUiComponent, selectors: [["basis-block-ui"]], decls: 3, vars: 1, consts: [[3, "blocked"], [1, "center-screen"]], template: function BlockUiComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelementStart(0, "p-blockUI", 0);
        ɵngcc0.ɵɵelementStart(1, "div", 1);
        ɵngcc0.ɵɵelement(2, "p-progressSpinner");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        ɵngcc0.ɵɵproperty("blocked", ctx.isBlocked());
    } }, directives: [ɵngcc2.BlockUI, ɵngcc3.ProgressSpinner], styles: [".center-screen[_ngcontent-%COMP%]{display:flex;flex-direction:column;justify-content:center;align-items:center;text-align:center;min-height:100vh}"] });
BlockUiComponent.ctorParameters = () => [
    { type: BlockUiService }
];
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(BlockUiComponent, [{
        type: Component,
        args: [{
                selector: 'basis-block-ui',
                template: "<p-blockUI [blocked]=\"isBlocked()\">\n    <div class=\"center-screen\">\n        <p-progressSpinner></p-progressSpinner>\n    </div>\n</p-blockUI>\n",
                styles: [".center-screen{display:flex;flex-direction:column;justify-content:center;align-items:center;text-align:center;min-height:100vh}"]
            }]
    }], function () { return [{ type: ɵngcc1.BlockUiService }]; }, null); })();
