// ng-block-ui/esm2015/router/block-ui-prevent-navigation.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { Injectable } from '@angular/core';
import { BlockUIService, BLOCKUI_DEFAULT } from 'ng-block-ui';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from 'ng-block-ui';
export class BlockUIPreventNavigation {
    /**
     * @param {?} blockUIService
     */
    constructor(blockUIService) {
        this.blockUIService = blockUIService;
    }
    /**
     * @return {?}
     */
    canActivate() {
        return !this.blockUIService.isActive(BLOCKUI_DEFAULT);
    }
    /**
     * @return {?}
     */
    canActivateChild() {
        return !this.blockUIService.isActive(BLOCKUI_DEFAULT);
    }
}
BlockUIPreventNavigation.ɵfac = function BlockUIPreventNavigation_Factory(t) { return new (t || BlockUIPreventNavigation)(ɵngcc0.ɵɵinject(ɵngcc1.BlockUIService)); };
BlockUIPreventNavigation.ɵprov = ɵngcc0.ɵɵdefineInjectable({ token: BlockUIPreventNavigation, factory: BlockUIPreventNavigation.ɵfac });
/** @nocollapse */
BlockUIPreventNavigation.ctorParameters = () => [
    { type: BlockUIService }
];
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(BlockUIPreventNavigation, [{
        type: Injectable
    }], function () { return [{ type: ɵngcc1.BlockUIService }]; }, null); })();
function BlockUIPreventNavigation_tsickle_Closure_declarations() {
    /** @type {?} */
    BlockUIPreventNavigation.prototype.blockUIService;
}
